<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Analyze and Integrate Any Type of Nucleotide Recoding RNA-Seq Data • EZbakR</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Analyze and Integrate Any Type of Nucleotide Recoding RNA-Seq Data">
<meta name="description" content="A complete rewrite and reimagining of bakR (see Vock et al. (2025) &lt;doi:10.1371/journal.pcbi.1013179&gt;). Designed to support a wide array of analyses of nucleotide recoding RNA-seq (NR-seq) datasets of any type, including TimeLapse-seq/SLAM-seq/TUC-seq, Start-TimeLapse-seq (STL-seq), TT-TimeLapse-seq (TT-TL-seq), and subcellular NR-seq. EZbakR extends standard NR-seq standard NR-seq mutational modeling to support multi-label analyses (e.g., 4sU and 6sG dual labeling), and implements an improved hierarchical model to better account for transcript-to-transcript variance in metabolic label incorporation. EZbakR also generalized dynamical systems modeling of NR-seq data to support analyses of premature mRNA processing and flow between subcellular compartments. Finally, EZbakR implements flexible and well-powered comparative analyses of all estimated parameters via design matrix-specified generalized linear modeling.">
<meta property="og:description" content="A complete rewrite and reimagining of bakR (see Vock et al. (2025) &lt;doi:10.1371/journal.pcbi.1013179&gt;). Designed to support a wide array of analyses of nucleotide recoding RNA-seq (NR-seq) datasets of any type, including TimeLapse-seq/SLAM-seq/TUC-seq, Start-TimeLapse-seq (STL-seq), TT-TimeLapse-seq (TT-TL-seq), and subcellular NR-seq. EZbakR extends standard NR-seq standard NR-seq mutational modeling to support multi-label analyses (e.g., 4sU and 6sG dual labeling), and implements an improved hierarchical model to better account for transcript-to-transcript variance in metabolic label incorporation. EZbakR also generalized dynamical systems modeling of NR-seq data to support analyses of premature mRNA processing and flow between subcellular compartments. Finally, EZbakR implements flexible and well-powered comparative analyses of all estimated parameters via design matrix-specified generalized linear modeling.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">EZbakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/Quickstart.html">EZbakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="articles/EZget.html">Convenient fetching of EZbakR outputs: EZget()</a></li>
    <li><a class="dropdown-item" href="articles/EstimateFractions.html">Universal first step: EstimateFractions()</a></li>
    <li><a class="dropdown-item" href="articles/EstimateKinetics.html">Standard kinetic parameter estimation: EstimateKinetics()</a></li>
    <li><a class="dropdown-item" href="articles/Linear-modeling.html">Linear modeling in EZbakR</a></li>
    <li><a class="dropdown-item" href="articles/EZDynamics.html">Dynamical systems modeling in EZbakR: EZDynamics()</a></li>
    <li><a class="dropdown-item" href="articles/EZQC.html">QCing your NR-seq data with EZbakR: EZQC()</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/isaacvock/EZbakR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="welcome-to-ezbakr">Welcome to EZbakR!<a class="anchor" aria-label="anchor" href="#welcome-to-ezbakr"></a>
</h1></div>
<p>EZbakR is a highly flexible tool for analyses of nucleotide recoding RNA-seq datasets (NR-seq; e.g., <a href="https://www.nature.com/articles/nmeth.4582" class="external-link">TimeLapse-seq</a>, <a href="https://www.nature.com/articles/nmeth.4435" class="external-link">SLAM-seq</a>, <a href="https://pubmed.ncbi.nlm.nih.gov/31768978/" class="external-link">TUC-seq</a>, etc.). See <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1013179" class="external-link">our paper</a> for a discussion of the motivation behind EZbakR and its companion pipeline <a href="https://github.com/isaacvock/fastq2EZbakR" class="external-link">fastq2EZbakR</a>, as well as validation of all of its novel functionality.</p>
<p><a href="https://isaacvock.github.io/posts/ezbakR_update/" class="external-link">EZbakR is now on CRAN</a>, so to install, you can now run:</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'EZbakR'</span><span class="op">)</span></span></code></pre>
<p>To install the up-to-date development version, you can still run:</p>
<pre><code><span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://roxygen2.r-lib.org/" class="external-link">"roxygen2"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"roxygen2"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://remotes.r-lib.org" class="external-link">"remotes"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"isaacvock/EZbakR"</span><span class="op">)</span></span></code></pre>
<p>Documentation is here: <a href="https://isaacvock.github.io/EZbakR/">https://isaacvock.github.io/EZbakR/</a></p>
<div class="section level3">
<h3 id="want-to-learn-more-about-how-to-interpret-and-analyze-nr-seq-data">Want to learn more about how to interpret and analyze NR-seq data?<a class="anchor" aria-label="anchor" href="#want-to-learn-more-about-how-to-interpret-and-analyze-nr-seq-data"></a>
</h3>
<p>Check out some of the blogs I have written on the topic:</p>
<ol style="list-style-type: decimal">
<li>
<a href="https://isaacvock.github.io/posts/post-with-code/" class="external-link">An introduction on how to interpret NR-seq data</a>. In this post, I use the task of building up an NR-seq simulator in R to help you develop an intuition for the important features of NR-seq data. Code is interactive and can be edited and re-executed in your browser without any additional setup (thanks to Quarto-live and webr).</li>
<li>
<a href="https://isaacvock.github.io/posts/nrseq_analysis/" class="external-link">An introduction on how to analyze NR-seq data</a>. In this post, I motivate the concept of mixture modeling, the gold standard analysis strategy implemented in tools like EZbakR and GRAND-SLAM. I also show you how to implement the strategy yourself in R (code is interactive once again).</li>
<li>
<a href="https://isaacvock.github.io/posts/nrseq_deepdive/" class="external-link">An in-depth NR-seq review</a>: Covers the history, data interpretation, data processing, analyzing processed data, designing experiments, and extensions of NR-seq that have been developed. This is taken from Chapter 1 of my <a href="https://github.com/isaacvock/Thesis" class="external-link">thesis</a>.</li>
</ol>
</div>
<div class="section level2">
<h2 id="vignettes">Vignettes<a class="anchor" aria-label="anchor" href="#vignettes"></a>
</h2>
<p>Currently, the following functionalities have dedicated vignettes:</p>
<ol style="list-style-type: decimal">
<li>
<a href="https://isaacvock.github.io/EZbakR/articles/Quickstart.html">Quickstart</a>: Takes you through the standard workflow, similar to bakR’s one and only workflow.</li>
<li>
<a href="https://isaacvock.github.io/EZbakR/articles/EstimateFractions.html">Estimating fractions</a>: Estimating the fraction of reads from each mutational population in your data. This is the nearly universal first step in all NR-seq analyses. This is done with EZbakR’s <a href="https://isaacvock.github.io/EZbakR/reference/EstimateFractions.html"><code>EstimateFractions()</code></a> function.</li>
<li>
<a href="https://isaacvock.github.io/EZbakR/articles/EstimateKinetics.html">Estimating kinetics</a>: Estimating kinetic parameters of synthesis and degradation in a standard NR-seq experiment. For standard, single label, NR-seq analyses, this is the next step in your analysis workflow after estimating the fraction of reads that are from labeled RNA. This is done with EZbakR’s <a href="https://isaacvock.github.io/EZbakR/reference/EstimateKinetics.html"><code>EstimateKinetics()</code></a> function.</li>
<li>
<a href="https://isaacvock.github.io/EZbakR/articles/EZQC.html">Quality Control</a>: Assessing the quality of your NR-seq data. This is done with EZbakR’s <a href="https://isaacvock.github.io/EZbakR/reference/EZQC.html"><code>EZQC()</code></a> function.</li>
<li>
<a href="https://isaacvock.github.io/EZbakR/articles/Linear-modeling.html">Comparative analyses</a>: Fitting a flexible generalized linear model to your NR-seq data so as to perform comparative analyses of estimated kinetic parameters that complements differential expression analyses. This is done with EZbakR’s <a href="https://isaacvock.github.io/EZbakR/reference/AverageAndRegularize.html"><code>AverageAndRegularize()</code></a> and <a href="https://isaacvock.github.io/EZbakR/reference/CompareParameters.html"><code>CompareParameters()</code></a> functions.</li>
<li>
<a href="https://isaacvock.github.io/EZbakR/articles/EZDynamics.html">Dynamical systems modeling</a>. For analyses of subcellular fractionation and/or pre-mRNA processing dynamics. This is done with EZbakR’s <a href="https://isaacvock.github.io/EZbakR/reference/EZDynamics.html"><code>EZDynamics()</code></a> function.</li>
<li>
<a href="https://isaacvock.github.io/EZbakR/articles/EZget.html">Navigating EZbakR output</a>. Conveniently fetching data from EZbakR analyses. This is done with EZbakR’s <a href="https://isaacvock.github.io/EZbakR/reference/EZget.html"><code>EZget()</code></a> function.</li>
</ol>
<p>Other implemented functionality that may be of interest includes:</p>
<ol style="list-style-type: decimal">
<li>Providing fractions or kinetic parameter estimates as input. The former works similarly to how it did in bakR, and is implemented via the <a href="https://isaacvock.github.io/EZbakR/reference/EZbakRFractions.html"><code>EZbakRFractions()</code></a> function. THe latter is unique to EZbakR and is implemented via the <a href="https://isaacvock.github.io/EZbakR/reference/EZbakRKinetics.html"><code>EZbakRKinetics()</code></a> function.</li>
<li>Simulating NR-seq data. There are a number of simulation functions implemented in EZbakR. <a href="https://isaacvock.github.io/EZbakR/reference/EZSimulate.html"><code>EZSimulate()</code></a> is a convenient wrapper to several of these.</li>
</ol>
</div>
<div class="section level2">
<h2 id="update-3152025-analyses-of-transcript-isoforms">Update (3/15/2025): Analyses of transcript isoforms<a class="anchor" aria-label="anchor" href="#update-3152025-analyses-of-transcript-isoforms"></a>
</h2>
<p>We recently (3/14/2025) put out <a href="https://www.biorxiv.org/content/10.1101/2025.03.12.642874v1" class="external-link">a preprint</a> describing a method by which to analyze the kinetics of individual transcript isoforms using short read NR-seq data from total RNA. While this strategy is touched on a little bit in one of the EZbakR vignettes (<a href="https://isaacvock.github.io/EZbakR/articles/EstimateFractions.html#isoform-deconvolution">this one</a>), I have also developed a full fastq-to-volcano plot walkthrough using real downsampled fastq files from that preprint so you can see how every step of the fastq2EZbakR and EZbakR pipeline needs to be configured/run for these analyses. The tutorial is <a href="https://isaacvock.github.io/Isoform_Tutorial_Docs/" class="external-link">here</a>, and the data used in that tutorial is <a href="https://github.com/isaacvock/Isoform_Analysis_Tutorial" class="external-link">here</a>. Over the next couple weeks I will be adding some extra details/analyses to this tutorial, but in its current form (as of 3/15/2025), all of the basics of performing isoform-level analyses are covered there. It also acts as a hand-on tutorial for all of the EZbakR-suite and can thus useful to checkout and try out even if you aren’t interested in this particular analysis strategy.</p>
</div>
<div class="section level2">
<h2 id="whats-new">What’s new?<a class="anchor" aria-label="anchor" href="#whats-new"></a>
</h2>
<p>EZbakR represents a complete rewrite of <a href="https://github.com/simonlabcode/bakR" class="external-link">bakR</a>. Improvements implemented in EZbakR include:</p>
<ol style="list-style-type: decimal">
<li>Modular function design that facilitates using EZbakR with any kind of NR-seq data, regardless of the experimental design or data details.</li>
<li>Extended mixture modeling capabilities. Includes:
<ul>
<li>Support for multi-label analyses.</li>
<li>Hierarchical mutation rate estimation strategy to allow for feature-specific mutation rates.</li>
<li>More efficient and accurate uncertainty quantification.</li>
</ul>
</li>
<li>Additional kinetic parameter estimation strategies:
<ul>
<li>Non-steady-state analyses as introduced in <a href="https://www.sciencedirect.com/science/article/pii/S1097276521004962" class="external-link">Narain et al., 2021</a>.</li>
<li>Short-feed analyses that assume negligible degradation of existing RNA.</li>
<li>Synthesis rate estimation is implemented as a part of all strategies.</li>
</ul>
</li>
<li>Improved uncertainty propogation so as to achieve performance of bakR’s slower implementations (Hybrid and MCMC) with a strategy as efficient as bakR’s most efficent implementation (MLE).</li>
<li>Removal of Stan dependencies. I love Stan, but having it as an R package dependency makes installation and maintenace more difficult.</li>
<li>Optional <a href="https://arrow.apache.org/" class="external-link">Apache Arrow</a> backend to help with analyses of larger-than-RAM datasets</li>
<li>Linear model-based averaging of replicate data to support more complex experimental designs and maximally flexible comparative analyses.</li>
<li>Greater flexibility in terms of the input data structure. Namely, multiple different features can be specified in your input cB table, and multiple different experimental details can be included in your input metadf table.</li>
<li>A novel transcript isoform deconvolution strategy that allows for isoform-specific kinetic parameter estimation.</li>
<li>Generalized linear dynamical systems modeling of NR-seq data. Supports analyses of subcellular fractionation NR-seq extensions, such as those described <a href="https://pubmed.ncbi.nlm.nih.gov/38964322/" class="external-link">here</a>, <a href="https://pubmed.ncbi.nlm.nih.gov/39548324/" class="external-link">here</a>, and <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1012059" class="external-link">here</a>. Also supports analyses of pre-mRNA processing dynamics.</li>
</ol>
<p>In the near future, EZbakR will support anything bakR can do that isn’t currently implemented (Namely <code>DissectMechanisms()</code> and various visualization functions). There are also a number of exciting developments on the horizon, so stay tuned!</p>
</div>
<div class="section level2">
<h2 id="what-is-nr-seq">What is NR-seq?<a class="anchor" aria-label="anchor" href="#what-is-nr-seq"></a>
</h2>
<p>NR-seq refers to a class of methods that combine RNA-seq, metabolic labeling, and unique metabolic label recoding chemistries. These methods were originally developed to dissect the kinetics of RNA synthesis and degradation. Excitingly though, a treasure trove of extensions of the original methods have been created over the years. To-date, nucleotide recoding has been combined with the likes of <a href="https://www.nature.com/articles/nmeth.4582" class="external-link">TT-seq</a>, <a href="https://www.sciencedirect.com/science/article/pii/S1097276521006869?via%3Dihub" class="external-link">Start-seq</a>, <a href="https://www.nature.com/articles/s41592-021-01250-z" class="external-link">Ribo-seq</a>, <a href="https://www.nature.com/articles/s41586-019-1369-y" class="external-link">scRNA-seq</a> (other examples of this <a href="https://www.nature.com/articles/s41592-020-0935-4" class="external-link">here</a>, <a href="https://www.nature.com/articles/s41587-020-0480-9" class="external-link">here</a>, and <a href="https://www.sciencedirect.com/science/article/pii/S2405471224001194" class="external-link">here</a>), <a href="https://www.nature.com/articles/s41587-023-01948-9" class="external-link">Perturb-seq</a>, <a href="https://www.biorxiv.org/content/10.1101/2020.05.01.073296v1" class="external-link">long-read sequencing</a>, and <a href="https://pubmed.ncbi.nlm.nih.gov/38964322/" class="external-link">subcellular fractionation</a>. In addition, while the original methods used 4-thiouridine (s<sup>4</sup>U), the same chemistry has been found to work with <a href="https://pubs.acs.org/doi/full/10.1021/jacs.8b08554" class="external-link">6-thioguanosine</a> (s<sup>6</sup>G), opening the door to dual-labeling experimental designs (e.g., <a href="https://pubmed.ncbi.nlm.nih.gov/36018791/" class="external-link">TILAC</a>). EZbakR and its companion pipeline <a href="https://github.com/isaacvock/fastq2EZbakR" class="external-link">fastq2EZbakR</a> aim to provide an integrated and flexible framework to support this exciting class of methods.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=EZbakR" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/isaacvock/EZbakR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/isaacvock/EZbakR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing EZbakR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Isaac Vock <br><small class="roles"> Author, maintainer, copyright holder </small> <a href="https://orcid.org/0000-0002-7178-6886" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
