<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get contrasts of estimated parameters — CompareParameters • EZbakR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get contrasts of estimated parameters — CompareParameters"><meta name="description" content="CompareParameters() calculates differences in parameters estimated by
AverageAndRegularize() or EZDynamics() and performs null hypothesis
statistical testing, comparing their values to a null hypothesis of 0."><meta property="og:description" content="CompareParameters() calculates differences in parameters estimated by
AverageAndRegularize() or EZDynamics() and performs null hypothesis
statistical testing, comparing their values to a null hypothesis of 0."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">EZbakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Quickstart.html">EZbakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="../articles/EZget.html">Convenient fetching of EZbakR outputs: EZget()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateFractions.html">Universal first step: EstimateFractions()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateKinetics.html">Standard kinetic parameter estimation: EstimateKinetics()</a></li>
    <li><a class="dropdown-item" href="../articles/Linear-modeling.html">Linear modeling in EZbakR</a></li>
    <li><a class="dropdown-item" href="../articles/EZDynamics.html">Dynamical systems modeling in EZbakR: EZDynamics()</a></li>
    <li><a class="dropdown-item" href="../articles/EZQC.html">QCing your NR-seq data with EZbakR: EZQC()</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/isaacvock/EZbakR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get contrasts of estimated parameters</h1>
      <small class="dont-index">Source: <a href="https://github.com/isaacvock/EZbakR/blob/main/R/Comparisons.R" class="external-link"><code>R/Comparisons.R</code></a></small>
      <div class="d-none name"><code>CompareParameters.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>CompareParameters()</code> calculates differences in parameters estimated by
<code><a href="AverageAndRegularize.html">AverageAndRegularize()</a></code> or <code><a href="EZDynamics.html">EZDynamics()</a></code> and performs null hypothesis
statistical testing, comparing their values to a null hypothesis of 0.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CompareParameters</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  <span class="va">design_factor</span>,</span>
<span>  <span class="va">reference</span>,</span>
<span>  <span class="va">experimental</span>,</span>
<span>  <span class="va">param_name</span>,</span>
<span>  parameter <span class="op">=</span> <span class="st">"log_kdeg"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"averages"</span>,</span>
<span>  <span class="va">param_function</span>,</span>
<span>  condition <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exactMatch <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  repeatID <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  formula_mean <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sd_grouping_factors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fit_params <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalize_by_median <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reference_levels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  experimental_levels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-obj">obj<a class="anchor" aria-label="anchor" href="#arg-obj"></a></dt>
<dd><p>An <code>EZbakRFit</code> object, which is an <code>EZbakRFractions</code> object on
which <code><a href="AverageAndRegularize.html">AverageAndRegularize()</a></code> has been run.</p></dd>


<dt id="arg-design-factor">design_factor<a class="anchor" aria-label="anchor" href="#arg-design-factor"></a></dt>
<dd><p>Name of factor from <code>metadf</code> whose parameter estimates at
different factor values you would like to compare. If you specify this, you need
to also specify <code>reference</code> and <code>experimental</code>. If <code>type</code> == "dynamics", this
can have multiple values, being the names of all of the factors you would like
to stratify a group by.</p></dd>


<dt id="arg-reference">reference<a class="anchor" aria-label="anchor" href="#arg-reference"></a></dt>
<dd><p>Name of reference <code>design_factor</code> factor level value. Difference
will be calculated as <code>experimental</code> - <code>reference</code>. If type == "dynamics", then this should specify the levels
of all of the <code>design_factor</code>(s) reference group. For example, if you have
multiple <code>design_factor</code>'s, then <code>reference</code> must be a named character vector with one element
per <code>design_factor</code>, with elements named the corresponding <code>design_factor</code>. For
example, if <code>design_factor</code> is c("genotype", "treatment"), and you would like
to compare genotype = "WT" and treatment = "untreated" (reference) to genotype = "KO" and
treatment = "treated", then <code>reference</code> would need to be a vector with
one element named "genotype", equal to "WT" and one element named "treatment"
equal to "untreated" (this example could be created with, <code>c(genotype = "WT", treatment = "untreated")</code>).</p></dd>


<dt id="arg-experimental">experimental<a class="anchor" aria-label="anchor" href="#arg-experimental"></a></dt>
<dd><p>Name of <code>design_factor</code> factor level value to compare to reference.
Difference will be calculated as <code>experimental</code> - <code>reference</code>. If type == "dynamics", then this should specify the levels
of all of the <code>design_factor</code>(s) reference group. For example, if you have
multiple <code>design_factor</code>'s, then <code>experimental</code> must be a named character vector with one element
per <code>design_factor</code>, with elements named the corresponding <code>design_factor</code>. For
example, if <code>design_factor</code> is c("genotype", "treatment"), and you would like
to compare genotype = "WT" and treatment = "untreated" (reference) to genotype = "KO" and
treatment = "treated", then <code>experimental</code> would need to be a vector with
one element named "genotype", equal to "KO" and one element named "treatment"
equal to "treated" (this example could be created with, <code>c(genotype = "KO", treatment = "treated")</code>).</p></dd>


<dt id="arg-param-name">param_name<a class="anchor" aria-label="anchor" href="#arg-param-name"></a></dt>
<dd><p>If you want to assess the significance of a single parameter,
rather than the comparison of two parameters, specify that one parameter's name
here.</p></dd>


<dt id="arg-parameter">parameter<a class="anchor" aria-label="anchor" href="#arg-parameter"></a></dt>
<dd><p>Parameter to average across replicates of a given condition.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>Type of table to use. Can either be "averages" or "dynamics".</p></dd>


<dt id="arg-param-function">param_function<a class="anchor" aria-label="anchor" href="#arg-param-function"></a></dt>
<dd><p>NOT YET IMPLEMENTED. Will allow you to specify more complicated
functions of parameters when hypotheses you need to test are combinations of parameters
rather than individual parameters or simple differences in two parameters.</p></dd>


<dt id="arg-condition">condition<a class="anchor" aria-label="anchor" href="#arg-condition"></a></dt>
<dd><p>Same as <code>design_factor</code>, will be deprecated in favor of the
former in later release.</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Character vector of the set of features you want to stratify
reads by and estimate proportions of each RNA population. The default of "all"
will use all feature columns in the <code>obj</code>'s cB.</p></dd>


<dt id="arg-exactmatch">exactMatch<a class="anchor" aria-label="anchor" href="#arg-exactmatch"></a></dt>
<dd><p>If TRUE, then <code>features</code> and <code>populations</code> have to exactly match
those for a given fractions table for that table to be used. Means that you can't
specify a subset of features or populations by default, since this is TRUE
by default.</p></dd>


<dt id="arg-repeatid">repeatID<a class="anchor" aria-label="anchor" href="#arg-repeatid"></a></dt>
<dd><p>If multiple <code>averages</code> tables exist with the same metadata,
then this is the numerical index by which they are distinguished.</p></dd>


<dt id="arg-formula-mean">formula_mean<a class="anchor" aria-label="anchor" href="#arg-formula-mean"></a></dt>
<dd><p>An R formula object specifying how the <code>parameter</code> of interest
depends on the sample characteristics for the averages object you want to use.</p></dd>


<dt id="arg-sd-grouping-factors">sd_grouping_factors<a class="anchor" aria-label="anchor" href="#arg-sd-grouping-factors"></a></dt>
<dd><p>Metadf columns should data was grouped by when estimating
standard deviations across replicates for the averages object you want to use.</p></dd>


<dt id="arg-fit-params">fit_params<a class="anchor" aria-label="anchor" href="#arg-fit-params"></a></dt>
<dd><p>Character vector of parameter names in the averages object
you would like to use.</p></dd>


<dt id="arg-normalize-by-median">normalize_by_median<a class="anchor" aria-label="anchor" href="#arg-normalize-by-median"></a></dt>
<dd><p>If TRUE, then median difference will be set equal to 0.
This can account for global biases in parameter estimates due to things like differences
in effective label times. Does risk eliminating real signal though, so user discretion
is advised.</p></dd>


<dt id="arg-reference-levels">reference_levels<a class="anchor" aria-label="anchor" href="#arg-reference-levels"></a></dt>
<dd><p>Same as <code>reference</code>, but exclusively parsed in case of
<code>type</code> == "dynamics, included for backwards compatibility.</p></dd>


<dt id="arg-experimental-levels">experimental_levels<a class="anchor" aria-label="anchor" href="#arg-experimental-levels"></a></dt>
<dd><p>Same as <code>experimental</code>, but exclusively parsed in case of
<code>type</code> == "dynamics, included for backwards compatibility.</p></dd>


<dt id="arg-overwrite">overwrite<a class="anchor" aria-label="anchor" href="#arg-overwrite"></a></dt>
<dd><p>If TRUE, then identical output will be overwritten if it exists.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>EZbakRData</code> object with an additional "comparisons" table, detailing
the result of a comparison of a parameter estimate's valules across two different
conditions.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The EZbakR website has an extensive vignette walking through various use cases
and parameters you can compare with <code>CompareParameters()</code>: <a href="https://isaacvock.github.io/EZbakR/articles/Linear-modeling.html">vignette link</a>.</p>
<p>There are essentially 3 scenarios that <code>CompareParameters()</code> can handle:</p><ul><li><p>Pairwise comparisons: compare <code>reference</code> to <code>experimental</code> parameter
estimates of a specified <code>design_factor</code> from <code><a href="AverageAndRegularize.html">AverageAndRegularize()</a></code>. log(<code>experimental</code> / <code>reference</code>) is
the computed "difference" in this case.</p></li>
<li><p>Assess the value of a single parameter, which itself should represent a
difference between other parameters. The name of this parameter can be specified
via the <code>param_name</code> argument. This is useful for various interaction models
where some of the parameters of these models may represent things like "effect of
A on condition X".</p></li>
<li><p>Pairwise comparison of dynamical systems model parameter estimate: similar
to the first case listed above, but now when <code>type == "dynamics"</code>. <code>design_factor</code> can
now be a vector of all the <code>metadf</code> columns you stratified parameter estimates by.</p></li>
</ul><p>Eventually, a 4th option via the currently non-functional <code>param_function</code> argument
will be implemented, which will allow you to specify functions of parameters to be assessed,
which can be useful for certain interaction models.</p>
<p><code>CompareParameters()</code> calculates p-values using what is essentially an asymptotic Wald test,
meaning that a standard normal distribution is integrated. P-values are then multiple-test
adjusted using the method of Benjamini and Hochberg, implemented in R's <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust()</a></code>
function.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Simulate data to analyze</span></span></span>
<span class="r-in"><span><span class="va">simdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="EZSimulate.html">EZSimulate</a></span><span class="op">(</span><span class="fl">30</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create EZbakR input</span></span></span>
<span class="r-in"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="EZbakRData.html">EZbakRData</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span>, <span class="va">simdata</span><span class="op">$</span><span class="va">metadf</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Estimate Fractions</span></span></span>
<span class="r-in"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="EstimateFractions.html">EstimateFractions</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating mutation rates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Summarizing data for feature(s) of interest</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Averaging out the nucleotide counts for improved efficiency</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating fractions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing output</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Estimate Kinetics</span></span></span>
<span class="r-in"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="EstimateKinetics.html">EstimateKinetics</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Average estimates across replicate</span></span></span>
<span class="r-in"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="AverageAndRegularize.html">AverageAndRegularize</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Fitting linear model</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating coverage vs. variance trend</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Regularizing variance estimates</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Compare parameters across conditions</span></span></span>
<span class="r-in"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu">CompareParameters</span><span class="op">(</span></span></span>
<span class="r-in"><span><span class="va">ezbdo</span>,</span></span>
<span class="r-in"><span>design_factor <span class="op">=</span> <span class="st">"treatment"</span>,</span></span>
<span class="r-in"><span>reference <span class="op">=</span> <span class="st">"treatment1"</span>,</span></span>
<span class="r-in"><span>experimental <span class="op">=</span> <span class="st">"treatment2"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

