<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Correct for experimental/bioinformatic dropout of labeled RNA. — CorrectDropout • EZbakR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Correct for experimental/bioinformatic dropout of labeled RNA. — CorrectDropout"><meta name="description" content="Uses the strategy described here, and similar to that originally presented
in Berg et al. 2024."><meta property="og:description" content="Uses the strategy described here, and similar to that originally presented
in Berg et al. 2024."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">EZbakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Quickstart.html">EZbakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="../articles/EZget.html">Convenient fetching of EZbakR outputs: EZget()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateFractions.html">Universal first step: EstimateFractions()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateKinetics.html">Standard kinetic parameter estimation: EstimateKinetics()</a></li>
    <li><a class="dropdown-item" href="../articles/Linear-modeling.html">Linear modeling in EZbakR</a></li>
    <li><a class="dropdown-item" href="../articles/EZDynamics.html">Dynamical systems modeling in EZbakR: EZDynamics()</a></li>
    <li><a class="dropdown-item" href="../articles/EZQC.html">QCing your NR-seq data with EZbakR: EZQC()</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Correct for experimental/bioinformatic dropout of labeled RNA.</h1>

      <div class="d-none name"><code>CorrectDropout.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Uses the strategy described <a href="https://simonlabcode.github.io/bakR/articles/Dropout.html" class="external-link">here</a>, and similar to that originally presented
in <a href="https://academic.oup.com/nar/article/52/7/e35/7612100" class="external-link">Berg et al. 2024</a>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CorrectDropout</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  strategy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grandR"</span>, <span class="st">"bakR"</span><span class="op">)</span>,</span>
<span>  grouping_factors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  populations <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fraction_design <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  repeatID <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exactMatch <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  read_cutoff <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  dropout_cutoff <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-obj">obj<a class="anchor" aria-label="anchor" href="#arg-obj"></a></dt>
<dd><p>An EZbakRFractions object, which is an EZbakRData object on which
you have run <code><a href="EstimateFractions.html">EstimateFractions()</a></code>.</p></dd>


<dt id="arg-strategy">strategy<a class="anchor" aria-label="anchor" href="#arg-strategy"></a></dt>
<dd><p>Which dropout correction strategy to use. Options are:</p><ul><li><p>grandR: Described <a href="https://academic.oup.com/nar/article/52/7/e35/7612100" class="external-link">here</a>.
Cite that work and <a href="https://www.nature.com/articles/s41467-023-39163-4" class="external-link">grandR</a> if using this strategy. Quasi-non-parametric strategy
that finds an estimate of the dropout rate that eliminates any linear correlation
between the newness of a transcript and the difference in +s4U and -s4U normalized
read counts.</p></li>
<li><p>bakR: Described <a href="https://simonlabcode.github.io/bakR/articles/Dropout.html" class="external-link">here</a>.
Uses a simple generative model of dropout to derive a likelihood function, and the
dropout rate is estimated via the method of maximum likelihood.</p></li>
</ul><p>The "bakR" strategy has the advantage of being model-derived, making it possible
to assess model fit and thus whether the simple assumptions of both the "bakR"
and "grandR" dropout models are met. The "grandR" strategy has the advantage of
being more robust. Thus, the "grandR" strategy is currently used by default.</p></dd>


<dt id="arg-grouping-factors">grouping_factors<a class="anchor" aria-label="anchor" href="#arg-grouping-factors"></a></dt>
<dd><p>Which sample-detail columns in the metadf should be used
to group -s4U samples by for calculating the average -s4U RPM? The default value of
<code>NULL</code> will cause all sample-detail columns to be used.</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Character vector of the set of features you want to stratify
reads by and estimate proportions of each RNA population. The default of <code>NULL</code>
will expect there to be only one fractions table in the EZbakRFractions object.</p></dd>


<dt id="arg-populations">populations<a class="anchor" aria-label="anchor" href="#arg-populations"></a></dt>
<dd><p>Mutational populations that were analyzed to generate the
fractions table to use. For example, this would be "TC" for a standard
s4U-based nucleotide recoding experiment.</p></dd>


<dt id="arg-fraction-design">fraction_design<a class="anchor" aria-label="anchor" href="#arg-fraction-design"></a></dt>
<dd><p>"Design matrix" specifying which RNA populations exist
in your samples. By default, this will be created automatically and will assume
that all combinations of the <code>mutrate_populations</code> you have requested to analyze are
present in your data. If this is not the case for your data, then you will have
to create one manually. See docs for <code>EstimateFractions</code> (run ?EstimateFractions()) for more details.</p></dd>


<dt id="arg-repeatid">repeatID<a class="anchor" aria-label="anchor" href="#arg-repeatid"></a></dt>
<dd><p>If multiple <code>fractions</code> tables exist with the same metadata,
then this is the numerical index by which they are distinguished.</p></dd>


<dt id="arg-exactmatch">exactMatch<a class="anchor" aria-label="anchor" href="#arg-exactmatch"></a></dt>
<dd><p>If TRUE, then <code>features</code> must exactly match the <code>features</code>
metadata for a given fractions table for it to be used. Means that you cannot
specify a subset of features by default. Set this to FALSE if you would like
to specify a feature subset.</p></dd>


<dt id="arg-read-cutoff">read_cutoff<a class="anchor" aria-label="anchor" href="#arg-read-cutoff"></a></dt>
<dd><p>Minimum number of reads for a feature to be used to fit
the dropout model.</p></dd>


<dt id="arg-dropout-cutoff">dropout_cutoff<a class="anchor" aria-label="anchor" href="#arg-dropout-cutoff"></a></dt>
<dd><p>Maximum ratio of -s4U:+s4U RPMs for a feature to be
used to fit the dropout model (i.e., simple outlier filtering cutoff).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An <code>EZbakRData</code> object with the specified "fractions" table replaced
with a dropout corrected table.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Dropout is the disproportionate loss of labeled RNA/reads from said RNA
described independently <a href="https://academic.oup.com/nar/article/52/7/e35/7612100" class="external-link">here</a>
and <a href="https://www.biorxiv.org/content/10.1101/2023.05.24.542133v1" class="external-link">here</a>. It can originate from a combination of
bioinformatic (loss of high mutation content reads due to alignment problems),
technical (loss of labeled RNA during RNA extraction), and biological (transcriptional
shutoff in rare cases caused by metabolic label toxicity) sources.
<code>CorrectDropout()</code> compares label-fed and label-free controls from the same
experimental conditions to estimate and correct for this dropout. It assumes
that there is a single number (referred to as the dropout rate, or pdo) which
describes the rate at which labeled RNA is lost (relative to unlabeled RNA).
pdo ranges from 0 (no dropout) to 1 (complete loss of all labeled RNA), and
is thus interpreted as the percentage of labeled RNA/reads from labeled RNA
disproportionately lost, relative to the equivalent unlabeled species.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Simulate data to analyze</span></span></span>
<span class="r-in"><span><span class="va">simdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="EZSimulate.html">EZSimulate</a></span><span class="op">(</span><span class="fl">30</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create EZbakR input</span></span></span>
<span class="r-in"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="EZbakRData.html">EZbakRData</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span>, <span class="va">simdata</span><span class="op">$</span><span class="va">metadf</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Estimate Fractions</span></span></span>
<span class="r-in"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="EstimateFractions.html">EstimateFractions</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating mutation rates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Summarizing data for feature(s) of interest</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Averaging out the nucleotide counts for improved efficiency</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating fractions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing output</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Correct for dropout</span></span></span>
<span class="r-in"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu">CorrectDropout</span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimated rates of dropout are:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>    sample       pdo</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1 sample1 0.0100000</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2 sample2 0.0100000</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 sample3 0.0100000</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 4 sample4 0.3799101</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5 sample5 0.4072284</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 6 sample6 0.4034133</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

