<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Estimate isoform-specific fraction news (or more generally "fractions"). — EstimateIsoformFractions • EZbakR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Estimate isoform-specific fraction news (or more generally " fractions estimateisoformfractions><meta name="description" content="Combines the output of EstimateFractions with transcript isoform
quantification performed by an outside tool (e.g., RSEM, kallisto, salmon, etc.)
to infer transcript isoform-specific fraction news (or more generally fraction
of reads coming from a particular mutation population)."><meta property="og:description" content="Combines the output of EstimateFractions with transcript isoform
quantification performed by an outside tool (e.g., RSEM, kallisto, salmon, etc.)
to infer transcript isoform-specific fraction news (or more generally fraction
of reads coming from a particular mutation population)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">EZbakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Quickstart.html">EZbakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="../articles/EZget.html">EZget</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateFractions.html">EstimateFractions</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateKinetics.html">EstimateKinetics</a></li>
    <li><a class="dropdown-item" href="../articles/Linear-modeling.html">Linear modeling in EZbakR</a></li>
    <li><a class="dropdown-item" href="../articles/EZDynamics.html">Dynamical systems modeling in EZbakR</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Estimate isoform-specific fraction news (or more generally "fractions").</h1>

      <div class="d-none name"><code>EstimateIsoformFractions.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Combines the output of <code>EstimateFractions</code> with transcript isoform
quantification performed by an outside tool (e.g., RSEM, kallisto, salmon, etc.)
to infer transcript isoform-specific fraction news (or more generally fraction
of reads coming from a particular mutation population).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">EstimateIsoformFractions</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  populations <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fraction_design <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  repeatID <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exactMatch <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fraction_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  quant_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_to_transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  TPM_min <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  count_min <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-obj">obj<a class="anchor" aria-label="anchor" href="#arg-obj"></a></dt>
<dd><p>An <code>EZbakRData</code> object</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Character vector of the set of features you want to stratify
reads by and estimate proportions of each RNA population. The default of "all"
will use all feature columns in the <code>obj</code>'s cB.</p></dd>


<dt id="arg-populations">populations<a class="anchor" aria-label="anchor" href="#arg-populations"></a></dt>
<dd><p>Mutational populations that were analyzed to generate the
fractions table to use. For example, this would be "TC" for a standard
s4U-based nucleotide recoding experiment.</p></dd>


<dt id="arg-fraction-design">fraction_design<a class="anchor" aria-label="anchor" href="#arg-fraction-design"></a></dt>
<dd><p>"Design matrix" specifying which RNA populations exist
in your samples. By default, this will be created automatically and will assume
that all combinations of the <code>mutrate_populations</code> you have requested to analyze are
present in your data. If this is not the case for your data, then you will have
to create one manually. See docs for <code>EstimateFractions</code> (run ?EstimateFractions()) for more details.</p></dd>


<dt id="arg-repeatid">repeatID<a class="anchor" aria-label="anchor" href="#arg-repeatid"></a></dt>
<dd><p>If multiple <code>fractions</code> tables exist with the same metadata,
then this is the numerical index by which they are distinguished.</p></dd>


<dt id="arg-exactmatch">exactMatch<a class="anchor" aria-label="anchor" href="#arg-exactmatch"></a></dt>
<dd><p>If TRUE, then <code>features</code> and <code>populations</code> have to exactly match
those for a given fractions table for that table to be used. Means that you can't
specify a subset of features or populations by default, since this is TRUE
by default.</p></dd>


<dt id="arg-fraction-name">fraction_name<a class="anchor" aria-label="anchor" href="#arg-fraction-name"></a></dt>
<dd><p>Name of fraction estimate table to use. Should be stored in the
<code>obj$fractions</code> list under this name. Can also rely on specifying <code>features</code> and/or <code>populations</code>
and having <code><a href="EZget.html">EZget()</a></code> find it.</p></dd>


<dt id="arg-quant-name">quant_name<a class="anchor" aria-label="anchor" href="#arg-quant-name"></a></dt>
<dd><p>Name of transcript isoform quantification table to use. Should be stored
in the obj$readcounts list under this name. Use <code><a href="ImportIsoformQuant.html">ImportIsoformQuant()</a></code> to create
this table. If <code>quant_name</code> is <code>NULL</code>, it will search for tables containing the string
"isoform_quant" in their name, as that is the naming convention used by <code><a href="ImportIsoformQuant.html">ImportIsoformQuant()</a></code>.
If more than one such table exists, an error will be thrown and you will have to specify
the exact name in <code>quant_name</code>.</p></dd>


<dt id="arg-gene-to-transcript">gene_to_transcript<a class="anchor" aria-label="anchor" href="#arg-gene-to-transcript"></a></dt>
<dd><p>Table with columns <code>transcript_id</code> and all feature related
columns that appear in the relevant fractions table. This is only relevant as a hack to
to deal with the case where STAR includes in its transcriptome alignment transcripts
on the opposite strand from where the RNA actually originated. This table will be used
to filter out such transcript-feature combinations that should not exist.</p></dd>


<dt id="arg-overwrite">overwrite<a class="anchor" aria-label="anchor" href="#arg-overwrite"></a></dt>
<dd><p>If TRUE and a fractions estimate output already exists that
would possess the same metadata (features analyzed, populations analyzed,
and fraction_design), then it will get overwritten with the new output. Else,
it will be saved as a separate output with the same name + "_#" where "#" is a
numerical ID to distinguish the similar outputs.</p></dd>


<dt id="arg-tpm-min">TPM_min<a class="anchor" aria-label="anchor" href="#arg-tpm-min"></a></dt>
<dd><p>Minimum TPM for a transcript to be kept in analysis.</p></dd>


<dt id="arg-count-min">count_min<a class="anchor" aria-label="anchor" href="#arg-count-min"></a></dt>
<dd><p>Minimum expected_count for a transcript to be kept in analysis.</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

