<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>EZbakRData object helper function for users — EZbakRData • EZbakR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="EZbakRData object helper function for users — EZbakRData"><meta name="description" content="EZbakRData creates an object of class EZbakRData and checks the validity
of the provided input."><meta property="og:description" content="EZbakRData creates an object of class EZbakRData and checks the validity
of the provided input."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">EZbakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Quickstart.html">EZbakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="../articles/EZget.html">Convenient fetching of EZbakR outputs: EZget()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateFractions.html">Universal first step: EstimateFractions()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateKinetics.html">Standard kinetic parameter estimation: EstimateKinetics()</a></li>
    <li><a class="dropdown-item" href="../articles/Linear-modeling.html">Linear modeling in EZbakR</a></li>
    <li><a class="dropdown-item" href="../articles/EZDynamics.html">Dynamical systems modeling in EZbakR: EZDynamics()</a></li>
    <li><a class="dropdown-item" href="../articles/EZQC.html">QCing your NR-seq data with EZbakR: EZQC()</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1><code>EZbakRData</code> object helper function for users</h1>

      <div class="d-none name"><code>EZbakRData.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>EZbakRData</code> creates an object of class <code>EZbakRData</code> and checks the validity
of the provided input.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">EZbakRData</span><span class="op">(</span><span class="va">cB</span>, <span class="va">metadf</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-cb">cB<a class="anchor" aria-label="anchor" href="#arg-cb"></a></dt>
<dd><p>Data frame with the following columns:</p><ul><li><p>sample: Name given to particular sample from which data was collected.</p></li>
<li><p>mutational counts: Integers corresponding to the number of a particular
mutation seen in a sequencing read. The following column names are allowed:</p><ul><li><p>TC: Number of Thymine-to-Cytosine mutations</p></li>
<li><p>TA: Number of Thymine-to-Adenine mutations</p></li>
<li><p>TG: Number of Thymine-to-Guanine mutations</p></li>
<li><p>CT: Number of Cytosine-to-Thymine mutations</p></li>
<li><p>CA: Number of Cytosine-to-Adenine mutations</p></li>
<li><p>CG: Number of Cytosine-to-Guanine mutations</p></li>
<li><p>CU: Number of Cytosine-to-Uridine mutations</p></li>
<li><p>AT: Number of Adenine-to-Thymine mutations</p></li>
<li><p>AC: Number of Adenine-to-Cytosine mutations</p></li>
<li><p>AG: Number of Adenine-to-Guanine mutations</p></li>
<li><p>AU: Number of Adenine-to-Uridine mutations</p></li>
<li><p>GT: Number of Guanine-to-Thymine mutations</p></li>
<li><p>GC: Number of Guanine-to-Cytosine mutations</p></li>
<li><p>GA: Number of Guanine-to-Adenine mutations</p></li>
<li><p>GU: Number of Guanine-to-Uridine mutations</p></li>
<li><p>TN: Number of Thymine-to-Adenine/Cytosine/Guanine mutations</p></li>
<li><p>CN: Number of Cytosine-to-Adenine/Thymine/Guanine/Uridine mutations</p></li>
<li><p>AN: Number of Adenine-to-Thymine/Cytosine/Guanine/Uridine mutations</p></li>
<li><p>GN: Number of Guanine-to-Adenine/Cytosine/Thymine/Uridine mutations</p></li>
<li><p>UN: Number of Uridine-to-Adenine/Cytosine/Guanine mutations</p></li>
<li><p>NT: Number of Adenine/Cytosine/Guanine-to-Thymine mutations</p></li>
<li><p>NC: Number of Adenine/Thymine/Guanine/Uridine-to-Cytosine mutations</p></li>
<li><p>NtoA: Number of Thymine/Cytosine/Guanine/Uridine-to-Adenine mutations. (Naming convention changed because NA taken)</p></li>
<li><p>NU: Number of Cytosine/Guanine/Adenine-to-Uridine mutations.</p></li>
<li><p>NN: Number of any kind of mutation</p></li>
</ul></li>
<li><p>base nucleotide count: Integers corresponding to the number of instances of a
particular type of nucleotide whose mutations are tracked in a corresponding
mutation count column. The following column names are allowed:</p><ul><li><p>nT: Number of Thymines</p></li>
<li><p>nG: Number of Guanines</p></li>
<li><p>nA: Number of Adenines</p></li>
<li><p>nC: Number of Cytosines</p></li>
<li><p>nU: Number of Uridines</p></li>
<li><p>nN: number of any kind of nucleotide</p></li>
</ul></li>
<li><p>features: Any columns that cannot be interpreted as a mutation count
or base nucleotide count (and that aren't named <code>sample</code> or <code>n</code>) will be
interpreted as an ID for a genomic "feature" from which a read originated.
Common examples of features and typical column names for said features include:</p><ul><li><p>Genes; common column names: gene, gene_id, gene_name, GF</p></li>
<li><p>Genes-exonic; common column names: gene_exon, gene_id_exon, gene_name_exon, XF</p></li>
<li><p>Transcripts; common column names: transcripts, TF</p></li>
<li><p>Exonic bins; common column names: exonic_bins, EF, EB</p></li>
<li><p>Exons; common column names: exons, exon_ids</p></li>
</ul><p>In some cases, a read will often map to multiple features (e.g., exons). Many
functions in bakR expect each of the feature IDs in these cases to be separated
by <code>+</code>. For example, if a read overlaps with two exons, with IDs exon_1 and exon_2,
then the corresponding entry in a  column of exonic assignments would be "exon_1+exon_2".
The default expectation can be overwritten though and is thus not strictly enforced.</p></li>
<li><p>n: Number of reads with identical values for all other columns.</p></li>
</ul></dd>


<dt id="arg-metadf">metadf<a class="anchor" aria-label="anchor" href="#arg-metadf"></a></dt>
<dd><p>Data frame detailing various aspects of each of the samples included
in the cB. This includes:</p><ul><li><p><code>sample</code>: The sample ID, which should correspond to a sample ID in the provided cB.</p></li>
<li><p><code>tl</code>: Metabolic label time. There are several edge cases to be aware of:</p><ul><li><p>If more than one metabolic label was used in the set of samples described
by the metadf (e.g., s4U and s6G were used), then the <code>tl</code> column should be
replaced by <code>tl_&lt;muttype&gt;</code>, where <code>&lt;muttype&gt;</code> represents the corresponding mutation
type count column in the cB that the label whose incubation time will be listed
in this column. For example, if feeding with s4U in some samples and s6G in others,
then performing standard nucleotide recoding chemistry, you will include
<code>tl_TC</code> and <code>tl_GA</code> columns corresponding to the s4U and s6G label times, respectively.</p></li>
<li><p>If a pulse-chase experimental design was used (!!this is strongly discouraged
unless you have a legitimate reason to prefer this design to a pulse-label
design!!), then you should have columns named <code>tpulse</code> and <code>tchase</code>, corresponding
to the pulse and chase times respectively. The same _<muttype> convention should
be used in the case of multi-label pulse-chase designs.</muttype></p></li>
</ul></li>
<li><p>sample characteristics: The remaining columns can be named whatever you like
and should include distinguishing features of groups of samples. Common columns might
include:</p><ul><li><p><code>treatment</code>: The experimental treatment applied to a set of samples.
This could represent things like genetic knockouts or knockdowns, drug treatments, etc.</p></li>
<li><p><code>batch</code>: An ID for sets of samples that were collected and/or processed together.
Useful for regressing out technical batch effects</p></li>
</ul></li>
</ul></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An EZbakRData object. This is simply a list of the provide <code>cB</code> and
<code>metadf</code> with class <code>EZbakRData</code></p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

