<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Run quality control checks — EZQC • EZbakR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run quality control checks — EZQC"><meta name="description" content="EZQC() assesses multiple aspects of your NR-seq data and generates a number
of plots visualizing dataset-wide trends."><meta property="og:description" content="EZQC() assesses multiple aspects of your NR-seq data and generates a number
of plots visualizing dataset-wide trends."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">EZbakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Quickstart.html">EZbakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="../articles/EZget.html">Convenient fetching of EZbakR outputs: EZget()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateFractions.html">Universal first step: EstimateFractions()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateKinetics.html">Standard kinetic parameter estimation: EstimateKinetics()</a></li>
    <li><a class="dropdown-item" href="../articles/Linear-modeling.html">Linear modeling in EZbakR</a></li>
    <li><a class="dropdown-item" href="../articles/EZDynamics.html">Dynamical systems modeling in EZbakR: EZDynamics()</a></li>
    <li><a class="dropdown-item" href="../articles/EZQC.html">QCing your NR-seq data with EZbakR: EZQC()</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Run quality control checks</h1>

      <div class="d-none name"><code>EZQC.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>EZQC()</code> assesses multiple aspects of your NR-seq data and generates a number
of plots visualizing dataset-wide trends.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">EZQC</span><span class="op">(</span><span class="va">obj</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-obj">obj<a class="anchor" aria-label="anchor" href="#arg-obj"></a></dt>
<dd><p>EZbakRData or EZbakRFractions object.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Parameters passed to the class-specific method.
If you have provided an EZbakRFractions object, then these can be (all play the
same role as in <code><a href="EstimateKinetics.html">EstimateKinetics()</a></code>, that is they get passed to <code><a href="EZget.html">EZget()</a></code> to find
the fractions table you are interested in. See <code>?EstimateKinetics()</code> for details.):</p><ul><li><p><code>features</code></p></li>
<li><p><code>populations</code></p></li>
<li><p><code>fraction_design</code></p></li>
</ul><p>If you have provided an EZbakRData object, then these can be (all same the same
purpose as in <code>EstimateFractions</code>, so see <code>?EstimateFractions()</code> for details):</p><ul><li><p><code>mutrate_populations</code></p></li>
<li><p><code>features</code></p></li>
<li><p><code>filter_cols</code></p></li>
<li><p><code>filter_condition</code></p></li>
<li><p><code>remove_features</code></p></li>
</ul></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of <code>ggplot2</code> objects visualizing the various aspects of your data
assessed by <code>EZQC()</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>EZQC()</code> checks the following aspects of your NR-seq data. If you have passed
an <code>EZbakRData</code> object, then <code>EZQC()</code> checks:</p><ul><li><p>Raw mutation rates: In all sequencing reads, how many T's in the reference
were a C in the read? The hope is that raw mutation rates are higher than -label
controls in all +label samples. Higher raw mutation rates, especially when using
standard label times (e.g., 2 hours or more in mammalian systems), are typically
a sign of good label incorporation and low labeled RNA/read dropout. If you
don't have -label samples, know that background mutation rates are typically
less than 0.2%, so +label raw mutation rates several times higher than this
would be preferable.</p></li>
<li><p>Mutation rates in labeled and unlabeled reads: The raw mutation rate
counts all mutations in all reads. In a standard NR-seq experiment performed
with a single metabolic label, there are typically two populations of reads:</p><ol><li><p>Those from labeled RNA, having higher mutation rates due to chemical conversion/recoding
of the metabolic label and 2) those from unlabeled RNA, having lower, background
levels of mutations. <code>EZbakR</code> fits a two component mixture model to estimate the mutation
rates in these two populations separately. A successful NR-seq experiment should
have a labeled read mutation rate of &gt; 1% and a low background mutation rate of
&lt; 0.3%.</p></li>
<li><p>Read count replicate correlation: Simply the log10 read count correlation
for replicates, as inferred from your <code>metadf</code>.</p></li>
</ol></li>
</ul><p>If you have passed an <code>EZbakRFractions</code> object, i.e., the output of <code><a href="EstimateFractions.html">EstimateFractions()</a></code>,
then in addition to the checks in the <code>EZbakRData</code> input case, <code>EZQC()</code> also checks:</p><ul><li><p>Fraction labeled distribution: This is the distribution of feature-wise
fraction labeled's (or fraction high mutation content's) estimated by <code><a href="EstimateFractions.html">EstimateFractions()</a></code>.
The "ideal" is a distribution with mean around 0.5, as this maximizes the amount of RNA
with synthesis and degradation kinetics within the dynamic range of the experiment. In practice,
you will (and should) be at least a bit lower than this as longer label times risk
physiological impacts of metabolic labeling.</p></li>
<li><p>Fraction labeled replicate correlation: This is the logit(fraction labeled)
correlation between replicates, as inferred from your <code>metadf</code>.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Simulate data to analyze</span></span></span>
<span class="r-in"><span><span class="va">simdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="EZSimulate.html">EZSimulate</a></span><span class="op">(</span><span class="fl">30</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create EZbakR input</span></span></span>
<span class="r-in"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="EZbakRData.html">EZbakRData</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span>, <span class="va">simdata</span><span class="op">$</span><span class="va">metadf</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Estimate Fractions</span></span></span>
<span class="r-in"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="EstimateFractions.html">EstimateFractions</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating mutation rates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Summarizing data for feature(s) of interest</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Averaging out the nucleotide counts for improved efficiency</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating fractions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing output</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run QC</span></span></span>
<span class="r-in"><span><span class="va">QC</span> <span class="op">&lt;-</span> <span class="fu">EZQC</span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> CHECKING RAW MUTATION RATES...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> CHECKING INFERRED MUTATION RATES...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> CHECKING READ COUNT CORRELATIONS...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> log10(read counts) correlation for each pair of replicates are:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 12 × 3</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>    sample_1 sample_2 correlation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> sample1  sample2        0.984</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> sample1  sample3        0.983</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> sample1  sample7        0.986</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> sample2  sample3        0.979</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> sample2  sample7        0.986</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> sample3  sample7        0.983</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> sample4  sample5        0.988</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> sample4  sample6        0.987</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> sample4  sample8        0.990</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> sample5  sample6        0.991</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> sample5  sample8        0.991</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> sample6  sample8        0.990</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> log10(read counts) correlations are high, suggesting good reproducibility!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> CHECKING FRACTION LABELED DISTRIBUTIONS...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Average fractions for each sample are:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   sample  avg_fraction fraction_type  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> sample1        0.269 fraction_highTC</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> sample2        0.270 fraction_highTC</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> sample3        0.257 fraction_highTC</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> sample4        0.262 fraction_highTC</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> sample5        0.265 fraction_highTC</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> sample6        0.261 fraction_highTC</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Labeling rates (e.g., fraction labeled for single label experiments) look good!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> CHECKING FRACTION LABELED CORRELATIONS...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> logit(fraction_highTC) correlation for each pair of replicates are:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   sample_1 sample_2 correlation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> sample1  sample2        0.964</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> sample1  sample3        0.912</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> sample2  sample3        0.955</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> sample4  sample5        0.942</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> sample4  sample6        0.956</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> sample5  sample6        0.956</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> logit(fraction_highTC) correlations are high, suggesting good reproducibility!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

