<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Simulation of transcript isoform kinetic parameters. — SimulateIsoforms • EZbakR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulation of transcript isoform kinetic parameters. — SimulateIsoforms"><meta name="description" content="SimulateIsoforms() performs a simple simulation of isoform-specific kinetic
parameters to showcase and test EstimateIsoformFractions(). It assumes that
there are a set of reads (fraction of total set by funique parameter) which
map uniquely to a given isoform, while the rest are ambiguous to all isoforms
from that gene. Mutational content of these reads are simulated as in
SimulateOneRep()."><meta property="og:description" content="SimulateIsoforms() performs a simple simulation of isoform-specific kinetic
parameters to showcase and test EstimateIsoformFractions(). It assumes that
there are a set of reads (fraction of total set by funique parameter) which
map uniquely to a given isoform, while the rest are ambiguous to all isoforms
from that gene. Mutational content of these reads are simulated as in
SimulateOneRep()."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">EZbakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Quickstart.html">EZbakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="../articles/EZget.html">Convenient fetching of EZbakR outputs: EZget()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateFractions.html">Universal first step: EstimateFractions()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateKinetics.html">Standard kinetic parameter estimation: EstimateKinetics()</a></li>
    <li><a class="dropdown-item" href="../articles/Linear-modeling.html">Linear modeling in EZbakR</a></li>
    <li><a class="dropdown-item" href="../articles/EZDynamics.html">Dynamical systems modeling in EZbakR: EZDynamics()</a></li>
    <li><a class="dropdown-item" href="../articles/EZQC.html">QCing your NR-seq data with EZbakR: EZQC()</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Simulation of transcript isoform kinetic parameters.</h1>

      <div class="d-none name"><code>SimulateIsoforms.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>SimulateIsoforms()</code> performs a simple simulation of isoform-specific kinetic
parameters to showcase and test <code><a href="EstimateIsoformFractions.html">EstimateIsoformFractions()</a></code>. It assumes that
there are a set of reads (fraction of total set by <code>funique</code> parameter) which
map uniquely to a given isoform, while the rest are ambiguous to all isoforms
from that gene. Mutational content of these reads are simulated as in
<code><a href="SimulateOneRep.html">SimulateOneRep()</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">SimulateIsoforms</span><span class="op">(</span></span>
<span>  <span class="va">nfeatures</span>,</span>
<span>  nt <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seqdepth <span class="op">=</span> <span class="va">nfeatures</span> <span class="op">*</span> <span class="fl">2500</span>,</span>
<span>  label_time <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  sample_name <span class="op">=</span> <span class="st">"sampleA"</span>,</span>
<span>  feature_prefix <span class="op">=</span> <span class="st">"Gene"</span>,</span>
<span>  pnew <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  pold <span class="op">=</span> <span class="fl">0.002</span>,</span>
<span>  funique <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  readlength <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  Ucont <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  avg_numiso <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  psynthdiff <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  logkdeg_mean <span class="op">=</span> <span class="op">-</span><span class="fl">1.9</span>,</span>
<span>  logkdeg_sd <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  logksyn_mean <span class="op">=</span> <span class="fl">2.3</span>,</span>
<span>  logksyn_sd <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-nfeatures">nfeatures<a class="anchor" aria-label="anchor" href="#arg-nfeatures"></a></dt>
<dd><p>Number of "features" to simulate data for. Each feature will
have a simulated number of transcript isoforms</p></dd>


<dt id="arg-nt">nt<a class="anchor" aria-label="anchor" href="#arg-nt"></a></dt>
<dd><p>(Optional), can provide a vector of the number of isoforms you would
like to simulate for each of the <code>nfeatures</code> features. Vector can either be length
1, in which case that many isoforms will be simulated for all features, or length
equal to <code>nfeatures</code>.</p></dd>


<dt id="arg-seqdepth">seqdepth<a class="anchor" aria-label="anchor" href="#arg-seqdepth"></a></dt>
<dd><p>Total number of sequencing reads to simulate</p></dd>


<dt id="arg-label-time">label_time<a class="anchor" aria-label="anchor" href="#arg-label-time"></a></dt>
<dd><p>Length of s^4^U feed to simulate.</p></dd>


<dt id="arg-sample-name">sample_name<a class="anchor" aria-label="anchor" href="#arg-sample-name"></a></dt>
<dd><p>Character vector to assign to <code>sample</code> column of output simulated
data table (the cB table).</p></dd>


<dt id="arg-feature-prefix">feature_prefix<a class="anchor" aria-label="anchor" href="#arg-feature-prefix"></a></dt>
<dd><p>Name given to the i-th feature is <code>paste0(feature_prefix, i)</code>. Shows up in the
<code>feature</code> column of the output simulated data table.</p></dd>


<dt id="arg-pnew">pnew<a class="anchor" aria-label="anchor" href="#arg-pnew"></a></dt>
<dd><p>Probability that a T is mutated to a C if a read is new.</p></dd>


<dt id="arg-pold">pold<a class="anchor" aria-label="anchor" href="#arg-pold"></a></dt>
<dd><p>Probability that a T is mutated to a C if a read is old.</p></dd>


<dt id="arg-funique">funique<a class="anchor" aria-label="anchor" href="#arg-funique"></a></dt>
<dd><p>Fraction of reads that uniquely "map" to a single isoform.</p></dd>


<dt id="arg-readlength">readlength<a class="anchor" aria-label="anchor" href="#arg-readlength"></a></dt>
<dd><p>Length of simulated reads. In this simple simulation, all reads
are simulated as being exactly this length.</p></dd>


<dt id="arg-ucont">Ucont<a class="anchor" aria-label="anchor" href="#arg-ucont"></a></dt>
<dd><p>Probability that a nucleotide in a simulated read is a U.</p></dd>


<dt id="arg-avg-numiso">avg_numiso<a class="anchor" aria-label="anchor" href="#arg-avg-numiso"></a></dt>
<dd><p>Average number of isoforms for each feature. Feature-specific
isoform counts are drawn from a Poisson distribution with this average. NOTE:
to insure that all features have multiple isoforms, the simulated number of
isoforms drawn from a Poisson distribution is incremented by 2. Thus, the
actual average number of isoforms from each feature is <code>avg_numiso</code> + 2.</p></dd>


<dt id="arg-psynthdiff">psynthdiff<a class="anchor" aria-label="anchor" href="#arg-psynthdiff"></a></dt>
<dd><p>Percentage of genes for which all isoform abundance differences
are synthesis driven. If not synthesis driven, then isoform abundance differences
will be driven by differences in isoform kdegs.</p></dd>


<dt id="arg-logkdeg-mean">logkdeg_mean<a class="anchor" aria-label="anchor" href="#arg-logkdeg-mean"></a></dt>
<dd><p>meanlog of a log-normal distribution from which
kdegs are simulated</p></dd>


<dt id="arg-logkdeg-sd">logkdeg_sd<a class="anchor" aria-label="anchor" href="#arg-logkdeg-sd"></a></dt>
<dd><p>sdlog of a log-normal distribution from which
kdegs are simulated</p></dd>


<dt id="arg-logksyn-mean">logksyn_mean<a class="anchor" aria-label="anchor" href="#arg-logksyn-mean"></a></dt>
<dd><p>meanlog of a log-normal distribution from which
ksyns are simulated</p></dd>


<dt id="arg-logksyn-sd">logksyn_sd<a class="anchor" aria-label="anchor" href="#arg-logksyn-sd"></a></dt>
<dd><p>sdlog of a log-normal distribution from which
ksyns are simulated</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

