<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Simulate one replicate of multi-label NR-seq data — SimulateMultiLabel • EZbakR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate one replicate of multi-label NR-seq data — SimulateMultiLabel"><meta name="description" content="Generalizes SimulateOneRep() to simulate any combination of mutation types.
Currently, no kinetic model is used to relate certain parameters to the
fractions of reads belonging to each simulated mutational population. Instead
these fractions are drawn from a Dirichlet distribution with gene-specific
parameters."><meta property="og:description" content="Generalizes SimulateOneRep() to simulate any combination of mutation types.
Currently, no kinetic model is used to relate certain parameters to the
fractions of reads belonging to each simulated mutational population. Instead
these fractions are drawn from a Dirichlet distribution with gene-specific
parameters."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">EZbakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Quickstart.html">EZbakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="../articles/EZget.html">Convenient fetching of EZbakR outputs: EZget()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateFractions.html">Universal first step: EstimateFractions()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateKinetics.html">Standard kinetic parameter estimation: EstimateKinetics()</a></li>
    <li><a class="dropdown-item" href="../articles/Linear-modeling.html">Linear modeling in EZbakR</a></li>
    <li><a class="dropdown-item" href="../articles/EZDynamics.html">Dynamical systems modeling in EZbakR: EZDynamics()</a></li>
    <li><a class="dropdown-item" href="../articles/EZQC.html">QCing your NR-seq data with EZbakR: EZQC()</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/isaacvock/EZbakR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Simulate one replicate of multi-label NR-seq data</h1>
      <small class="dont-index">Source: <a href="https://github.com/isaacvock/EZbakR/blob/main/R/Simulations.R" class="external-link"><code>R/Simulations.R</code></a></small>
      <div class="d-none name"><code>SimulateMultiLabel.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generalizes SimulateOneRep() to simulate any combination of mutation types.
Currently, no kinetic model is used to relate certain parameters to the
fractions of reads belonging to each simulated mutational population. Instead
these fractions are drawn from a Dirichlet distribution with gene-specific
parameters.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">SimulateMultiLabel</span><span class="op">(</span></span>
<span>  <span class="va">nfeatures</span>,</span>
<span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TC"</span><span class="op">)</span>,</span>
<span>  fraction_design <span class="op">=</span> <span class="fu"><a href="create_fraction_design.html">create_fraction_design</a></span><span class="op">(</span><span class="va">populations</span><span class="op">)</span>,</span>
<span>  fractions_matrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  read_vect <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_name <span class="op">=</span> <span class="st">"sampleA"</span>,</span>
<span>  feature_prefix <span class="op">=</span> <span class="st">"Gene"</span>,</span>
<span>  kdeg_vect <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ksyn_vect <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  logkdeg_mean <span class="op">=</span> <span class="op">-</span><span class="fl">1.9</span>,</span>
<span>  logkdeg_sd <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  logksyn_mean <span class="op">=</span> <span class="fl">2.3</span>,</span>
<span>  logksyn_sd <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  phighs <span class="op">=</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.05</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">populations</span><span class="op">)</span><span class="op">)</span>, <span class="va">populations</span><span class="op">)</span>,</span>
<span>  plows <span class="op">=</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.002</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">populations</span><span class="op">)</span><span class="op">)</span>, <span class="va">populations</span><span class="op">)</span>,</span>
<span>  seqdepth <span class="op">=</span> <span class="va">nfeatures</span> <span class="op">*</span> <span class="fl">2500</span>,</span>
<span>  readlength <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  alpha_min <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  alpha_max <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  Ucont <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  Acont <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  Gcont <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  Ccont <span class="op">=</span> <span class="fl">0.25</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-nfeatures">nfeatures<a class="anchor" aria-label="anchor" href="#arg-nfeatures"></a></dt>
<dd><p>Number of "features" (e.g., genes) to simulate data for</p></dd>


<dt id="arg-populations">populations<a class="anchor" aria-label="anchor" href="#arg-populations"></a></dt>
<dd><p>Vector of mutation populations you want to simulate.</p></dd>


<dt id="arg-fraction-design">fraction_design<a class="anchor" aria-label="anchor" href="#arg-fraction-design"></a></dt>
<dd><p>Fraction design matrix, specifying which potential mutational
populations should actually exist. See ?EstimateFractions for more details.</p></dd>


<dt id="arg-fractions-matrix">fractions_matrix<a class="anchor" aria-label="anchor" href="#arg-fractions-matrix"></a></dt>
<dd><p>Matrix of fractions of each mutational population to simulate.
If not provided, this will be simulated. One row for each feature, one column for each
mutational population, rows should sum to 1.</p></dd>


<dt id="arg-read-vect">read_vect<a class="anchor" aria-label="anchor" href="#arg-read-vect"></a></dt>
<dd><p>Vector of length = <code>nfeatures</code>; specifies the number of reads
to be simulated for each feature. If this is not provided, the number of reads
simulated is equal to <code>round(seqdepth * (ksyn_i/kdeg_i)/sum(ksyn/kdeg))</code>. In other words,
the normalized steady-state abundance of a feature is multiplied by the total number
of reads to be simulated and rounded to the nearest integer.</p></dd>


<dt id="arg-sample-name">sample_name<a class="anchor" aria-label="anchor" href="#arg-sample-name"></a></dt>
<dd><p>Character vector to assign to <code>sample</code> column of output simulated
data table (the cB table).</p></dd>


<dt id="arg-feature-prefix">feature_prefix<a class="anchor" aria-label="anchor" href="#arg-feature-prefix"></a></dt>
<dd><p>Name given to the i-th feature is <code>paste0(feature_prefix, i)</code>. Shows up in the
<code>feature</code> column of the output simulated data table.</p></dd>


<dt id="arg-kdeg-vect">kdeg_vect<a class="anchor" aria-label="anchor" href="#arg-kdeg-vect"></a></dt>
<dd><p>Vector of length = <code>nfeatures</code>; specifies the degradation rate constant to use for each
feature's simulation. If this is not provided and <code>fn_vect</code> is, then <code>kdeg_vect = -log(1 - fn_vect)/label_time</code>.
If both <code>kdeg_vect</code> and <code>fn_vect</code> are not provided, each feature's <code>kdeg_vect</code> value is drawn from a log-normal distrubition
with meanlog = <code>logkdeg_mean</code> and sdlog = <code>logkdeg_sd</code>. <code>kdeg_vect</code> is actually only simulated in the case
where <code>read_vect</code> is also not provided, as it will be used to simulate read counts as described above.</p></dd>


<dt id="arg-ksyn-vect">ksyn_vect<a class="anchor" aria-label="anchor" href="#arg-ksyn-vect"></a></dt>
<dd><p>Vector of length = <code>nfeatures</code>; specifies the synthesis rate constant to use for each
feature's simulation. If this is not provided, and <code>read_vect</code> is also not provided, then each
feature's <code>ksyn_vect</code> value is drawn from a log-normal distribution with meanlog = <code>logksyn_mean</code> and
sdlog = <code>logksyn_sd</code>. ksyn's do not need to be simulated if <code>read_vect</code> is provided, as they only
influence read counts.</p></dd>


<dt id="arg-logkdeg-mean">logkdeg_mean<a class="anchor" aria-label="anchor" href="#arg-logkdeg-mean"></a></dt>
<dd><p>If necessary, meanlog of a log-normal distribution from which
kdegs are simulated</p></dd>


<dt id="arg-logkdeg-sd">logkdeg_sd<a class="anchor" aria-label="anchor" href="#arg-logkdeg-sd"></a></dt>
<dd><p>If necessary, sdlog of a log-normal distribution from which
kdegs are simulated</p></dd>


<dt id="arg-logksyn-mean">logksyn_mean<a class="anchor" aria-label="anchor" href="#arg-logksyn-mean"></a></dt>
<dd><p>If necessary, meanlog of a log-normal distribution from which
ksyns are simulated</p></dd>


<dt id="arg-logksyn-sd">logksyn_sd<a class="anchor" aria-label="anchor" href="#arg-logksyn-sd"></a></dt>
<dd><p>If necessary, sdlog of a log-normal distribution from which
ksyns are simulated</p></dd>


<dt id="arg-phighs">phighs<a class="anchor" aria-label="anchor" href="#arg-phighs"></a></dt>
<dd><p>Vector of probabilities of mutation rates in labeled reads of each type denoted in
<code>populations</code>. Should be a named vector, with names being the corresponding <code>population</code>.</p></dd>


<dt id="arg-plows">plows<a class="anchor" aria-label="anchor" href="#arg-plows"></a></dt>
<dd><p>Vector of probabilities of mutation rates in unlabeled reads of each type denoted in
<code>populations</code>. Should be a named vector, with names being the corresponding <code>population</code>.</p></dd>


<dt id="arg-seqdepth">seqdepth<a class="anchor" aria-label="anchor" href="#arg-seqdepth"></a></dt>
<dd><p>Only relevant if <code>read_vect</code> is not provided; in that case, this is
the total number of reads to simulate.</p></dd>


<dt id="arg-readlength">readlength<a class="anchor" aria-label="anchor" href="#arg-readlength"></a></dt>
<dd><p>Length of simulated reads. In this simple simulation, all reads
are simulated as being exactly this length.</p></dd>


<dt id="arg-alpha-min">alpha_min<a class="anchor" aria-label="anchor" href="#arg-alpha-min"></a></dt>
<dd><p>Minimum possible value of alpha element of Dirichlet random variable</p></dd>


<dt id="arg-alpha-max">alpha_max<a class="anchor" aria-label="anchor" href="#arg-alpha-max"></a></dt>
<dd><p>Maximum possible value of alpha element of Dirichlet random variable</p></dd>


<dt id="arg-ucont">Ucont<a class="anchor" aria-label="anchor" href="#arg-ucont"></a></dt>
<dd><p>Probability that a nucleotide in a simulated read is a U.</p></dd>


<dt id="arg-acont">Acont<a class="anchor" aria-label="anchor" href="#arg-acont"></a></dt>
<dd><p>Probability that a nucleotide in a simulated read is an A.</p></dd>


<dt id="arg-gcont">Gcont<a class="anchor" aria-label="anchor" href="#arg-gcont"></a></dt>
<dd><p>Probability that a nucleotide in a simulated read is a G.</p></dd>


<dt id="arg-ccont">Ccont<a class="anchor" aria-label="anchor" href="#arg-ccont"></a></dt>
<dd><p>Probability that a nucleotide in a simulated read is a C.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>List with two elements:</p><ul><li><p>cB: Tibble that can be passed as the <code>cB</code> arg to <code><a href="EZbakRData.html">EZbakRData()</a></code>.</p></li>
<li><p>ground_truth: Tibble containing simulated ground truth.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">simdata</span> <span class="op">&lt;-</span> <span class="fu">SimulateMultiLabel</span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The `x` argument of `as_tibble.matrix()` must have unique column names if</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> `.name_repair` is omitted as of tibble 2.0.0.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Using compatibility `.name_repair`.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">EZbakR</span> package.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/isaacvock/EZbakR/issues&gt;</span>.</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

