<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="SimulateIsoforms() performs a simple simulation of isoform-specific kinetic
parameters to showcase and test EstimateIsoformFractions(). It assumes that
there are a set of reads (fraction of total set by funique parameter) which
map uniquely to a given isoform, while the rest are ambiguous to all isoforms
from that gene. Mutational content of these reads are simulated as in
SimulateOneRep()."><title>Simulation of transcript isoform kinetic parameters. — SimulateIsoforms • EZbakR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulation of transcript isoform kinetic parameters. — SimulateIsoforms"><meta property="og:description" content="SimulateIsoforms() performs a simple simulation of isoform-specific kinetic
parameters to showcase and test EstimateIsoformFractions(). It assumes that
there are a set of reads (fraction of total set by funique parameter) which
map uniquely to a given isoform, while the rest are ambiguous to all isoforms
from that gene. Mutational content of these reads are simulated as in
SimulateOneRep()."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">EZbakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Quickstart.html">EZbakR for people in a hurry</a>
    <a class="dropdown-item" href="../articles/EZget.html">EZget</a>
    <a class="dropdown-item" href="../articles/EstimateFractions.html">EstimateFractions</a>
    <a class="dropdown-item" href="../articles/EstimateKinetics.html">EstimateKinetics</a>
    <a class="dropdown-item" href="../articles/AverageAndRegularize.html">AverageAndRegularize</a>
    <a class="dropdown-item" href="../articles/CompareParameters.html">CompareParameters</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Simulation of transcript isoform kinetic parameters.</h1>
      
      <div class="d-none name"><code>SimulateIsoforms.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>SimulateIsoforms()</code> performs a simple simulation of isoform-specific kinetic
parameters to showcase and test <code><a href="EstimateIsoformFractions.html">EstimateIsoformFractions()</a></code>. It assumes that
there are a set of reads (fraction of total set by <code>funique</code> parameter) which
map uniquely to a given isoform, while the rest are ambiguous to all isoforms
from that gene. Mutational content of these reads are simulated as in
<code><a href="SimulateOneRep.html">SimulateOneRep()</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">SimulateIsoforms</span><span class="op">(</span></span>
<span>  <span class="va">nfeatures</span>,</span>
<span>  nt <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seqdepth <span class="op">=</span> <span class="va">nfeatures</span> <span class="op">*</span> <span class="fl">2500</span>,</span>
<span>  label_time <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  sample_name <span class="op">=</span> <span class="st">"sampleA"</span>,</span>
<span>  feature_prefix <span class="op">=</span> <span class="st">"Gene"</span>,</span>
<span>  pnew <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  pold <span class="op">=</span> <span class="fl">0.002</span>,</span>
<span>  funique <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  readlength <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  Ucont <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  avg_numiso <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  psynthdiff <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  logkdeg_mean <span class="op">=</span> <span class="op">-</span><span class="fl">1.9</span>,</span>
<span>  logkdeg_sd <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  logksyn_mean <span class="op">=</span> <span class="fl">2.3</span>,</span>
<span>  logksyn_sd <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>nfeatures</dt>
<dd><p>Number of "features" to simulate data for. Each feature will
have a simulated number of transcript isoforms</p></dd>


<dt>nt</dt>
<dd><p>(Optional), can provide a vector of the number of isoforms you would
like to simulate for each of the <code>nfeatures</code> features. Vector can either be length
1, in which case that many isoforms will be simulated for all features, or length
equal to <code>nfeatures</code>.</p></dd>


<dt>seqdepth</dt>
<dd><p>Total number of sequencing reads to simulate</p></dd>


<dt>label_time</dt>
<dd><p>Length of s^4U feed to simulate.</p></dd>


<dt>sample_name</dt>
<dd><p>Character vector to assign to <code>sample</code> column of output simulated
data table (the cB table).</p></dd>


<dt>feature_prefix</dt>
<dd><p>Name given to the i-th feature is <code>paste0(feature_prefix, i)</code>. Shows up in the
<code>feature</code> column of the output simulated data table.</p></dd>


<dt>pnew</dt>
<dd><p>Probability that a T is mutated to a C if a read is new.</p></dd>


<dt>pold</dt>
<dd><p>Probability that a T is mutated to a C if a read is old.</p></dd>


<dt>funique</dt>
<dd><p>Fraction of reads that uniquely "map" to a single isoform.</p></dd>


<dt>readlength</dt>
<dd><p>Length of simulated reads. In this simple simulation, all reads
are simulated as being exactly this length.</p></dd>


<dt>Ucont</dt>
<dd><p>Probability that a nucleotide in a simulated read is a U.</p></dd>


<dt>avg_numiso</dt>
<dd><p>Average number of isoforms for each feature. Feature-specific
isoform counts are drawn from a Poisson distribution with this average. NOTE:
to insure that all features have multiple isoforms, the simulated number of
isoforms drawn from a Poisson distribution is incremented by 2. Thus, the
actual average number of isoforms from each feature is <code>avg_numiso</code> + 2.</p></dd>


<dt>psynthdiff</dt>
<dd><p>Percentage of genes for which all isoform abundance differences
are synthesis driven. If not synthesis driven, then isoform abundance differences
will be driven by differences in isoform kdegs.</p></dd>


<dt>logkdeg_mean</dt>
<dd><p>meanlog of a log-normal distribution from which
kdegs are simulated</p></dd>


<dt>logkdeg_sd</dt>
<dd><p>sdlog of a log-normal distribution from which
kdegs are simulated</p></dd>


<dt>logksyn_mean</dt>
<dd><p>meanlog of a log-normal distribution from which
ksyns are simulated</p></dd>


<dt>logksyn_sd</dt>
<dd><p>sdlog of a log-normal distribution from which
ksyns are simulated</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

