<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Dynamical systems modeling in EZbakR: EZDynamics() • EZbakR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Dynamical systems modeling in EZbakR: EZDynamics()">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">EZbakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Quickstart.html">EZbakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="../articles/EZget.html">Convenient fetching of EZbakR outputs: EZget()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateFractions.html">Universal first step: EstimateFractions()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateKinetics.html">Standard kinetic parameter estimation: EstimateKinetics()</a></li>
    <li><a class="dropdown-item" href="../articles/Linear-modeling.html">Linear modeling in EZbakR</a></li>
    <li><a class="dropdown-item" href="../articles/EZDynamics.html">Dynamical systems modeling in EZbakR: EZDynamics()</a></li>
    <li><a class="dropdown-item" href="../articles/EZQC.html">QCing your NR-seq data with EZbakR: EZQC()</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/isaacvock/EZbakR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Dynamical systems modeling in EZbakR: EZDynamics()</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/isaacvock/EZbakR/blob/main/vignettes/EZDynamics.Rmd" class="external-link"><code>vignettes/EZDynamics.Rmd</code></a></small>
      <div class="d-none name"><code>EZDynamics.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://isaacvock.github.io/EZbakR/">EZbakR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.stats.ox.ac.uk/pub/MASS4/" class="external-link">MASS</a></span><span class="op">)</span> <span class="co"># For plotting</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MASS'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     select</span></span>
<span></span>
<span><span class="co"># For coloring points by density:</span></span>
<span><span class="co"># Source: https://slowkow.com/notes/ggplot2-color-by-density/</span></span>
<span><span class="va">get_density</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">dens</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MASS/man/kde2d.html" class="external-link">kde2d</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">...</span><span class="op">)</span></span>
<span>  <span class="va">ix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/findInterval.html" class="external-link">findInterval</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">dens</span><span class="op">$</span><span class="va">x</span><span class="op">)</span></span>
<span>  <span class="va">iy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/findInterval.html" class="external-link">findInterval</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">dens</span><span class="op">$</span><span class="va">y</span><span class="op">)</span></span>
<span>  <span class="va">ii</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">ix</span>, <span class="va">iy</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">dens</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="va">ii</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Increasingly, subcellular fractionation is being combined with
NR-seq, allowing researchers to probe the kinetics of flow between
subcellular compartments. EZbakR’s <code><a href="../reference/EZDynamics.html">EZDynamics()</a></code> function
supports the analysis of this kind of data through generalized dynamical
systems modeling. In addition, <code>EZDyamics()</code>, when combined
with fastq2EZbakR’s flexible feature assignment strategy, opens the door
for analyses of both premature and mature RNA dynamics. It is even
possible to combine these two. In this vignette, we will learn how to
run <code><a href="../reference/EZDynamics.html">EZDynamics()</a></code>.</p>
</div>
<div class="section level2">
<h2 id="quickstart">Quickstart<a class="anchor" aria-label="anchor" href="#quickstart"></a>
</h2>
<p>Here we will quickly cover a couple use cases to efficiently showcase
the full extent of <code><a href="../reference/EZDynamics.html">EZDynamics()</a></code> functionality. In the
sections that follow, a more detailed description of everything shown
here will be provided.</p>
<div class="section level3">
<h3 id="modeling-cytoplasmic-to-nuclear-flow">Modeling cytoplasmic to nuclear flow<a class="anchor" aria-label="anchor" href="#modeling-cytoplasmic-to-nuclear-flow"></a>
</h3>
<p>Data: Whole cell, cytoplasmic fraction, and nuclear fraction NR-seq
data.</p>
<p>Goal: Fit the simple model 0 -&gt; N -&gt; C -&gt; 0; nuclear RNA (N)
is synthesized, exported to the cytoplasm (C), and degraded (0).</p>
<p>Pre-requisites:</p>
<ol style="list-style-type: decimal">
<li>Create EZbakRData object</li>
<li>Run EstimateFractions</li>
<li>Create table of feature lengths from annotation</li>
<li>Run AverageAndRegularize to average fractions across replicates</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### SIMULATE DATA</span></span>
<span></span>
<span><span class="va">simdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZSimulate.html">EZSimulate</a></span><span class="op">(</span>nfeatures <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                      ntreatments <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                      mode <span class="op">=</span> <span class="st">"dynamics"</span>,</span>
<span>                      label_time <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                      dynamics_preset <span class="op">=</span> <span class="st">"nuc2cyto"</span><span class="op">)</span></span>
<span>              </span>
<span><span class="co">### EZBAKR ANALYSES</span></span>
<span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZbakRData.html">EZbakRData</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span>, <span class="va">simdata</span><span class="op">$</span><span class="va">metadf</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EstimateFractions.html">EstimateFractions</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span>
<span><span class="co">#&gt; Estimating mutation rates</span></span>
<span><span class="co">#&gt; Summarizing data for feature(s) of interest</span></span>
<span><span class="co">#&gt; Averaging out the nucleotide counts for improved efficiency</span></span>
<span><span class="co">#&gt; Estimating fractions</span></span>
<span><span class="co">#&gt; Processing output</span></span>
<span></span>
<span></span>
<span><span class="co"># Lengths of features:</span></span>
<span><span class="co"># In simulation, all features are same length, so this is here for illustrative</span></span>
<span><span class="co"># purposes only</span></span>
<span><span class="va">feature_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  GF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span><span class="op">$</span><span class="va">GF</span><span class="op">)</span>,</span>
<span>  length <span class="op">=</span> <span class="fl">1000</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Averaging fractions rather than kinetics</span></span>
<span><span class="co"># Note the model, which specifies that we want to average data for a given</span></span>
<span><span class="co"># compartment AND label time (tl).</span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AverageAndRegularize.html">AverageAndRegularize</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                              formula_mean <span class="op">=</span> <span class="op">~</span><span class="va">tl</span><span class="op">:</span><span class="va">compartment</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>                              type <span class="op">=</span> <span class="st">"fractions"</span>,</span>
<span>                              feature_lengths <span class="op">=</span> <span class="va">feature_lengths</span>,</span>
<span>                              parameter <span class="op">=</span> <span class="st">"logit_fraction_highTC"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Fitting linear model</span></span>
<span><span class="co">#&gt; Estimating coverage vs. variance trend</span></span>
<span><span class="co">#&gt; Regularizing variance estimates</span></span></code></pre></div>
<p>Input to EZDynamics():</p>
<ol style="list-style-type: decimal">
<li>EZbakRData object on which you have run AverageAndRegularize</li>
<li>A matrix, referred to as a “graph”, describing the species you are
modeling and their relationships to one another:</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Input #2: the graph</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>,</span>
<span>                  <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">2</span>,</span>
<span>                  <span class="fl">3</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                nrow <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                ncol <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"N"</span>, <span class="st">"C"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Each row and column represents an RNA species being modeled</span></span>
<span><span class="co"># Here we are modeling the dynamics of nuclear (N) and cytoplasmic (C)</span></span>
<span><span class="co"># RNA. "0" represents no RNA, and must appear in the graph. N is synthesized</span></span>
<span><span class="co"># from 0 (i.e., its synthesis kinetics are 0-th order), and C is degraded to 0</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># Numbers in graph should range from 0 to the number of parameters in your model.</span></span>
<span><span class="co"># Order is arbitrary</span></span></code></pre></div>
<ol style="list-style-type: decimal">
<li>A list of equations describing how the RNA you actually measured is
related to the species you are modeling:</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">modeled_to_measured</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  nuclear <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">GF</span> <span class="op">~</span> <span class="va">N</span><span class="op">)</span>,</span>
<span>  cytoplasm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">GF</span> <span class="op">~</span> <span class="va">C</span><span class="op">)</span>,</span>
<span>  total <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">GF</span> <span class="op">~</span> <span class="va">C</span> <span class="op">+</span> <span class="va">N</span><span class="op">)</span> <span class="co"># total RNA is a combination of C and N</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Running EZDynamics:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># See documentation (?EZDynamics()) for descriptions of all parameters</span></span>
<span><span class="co"># specified. Bit of trickery here with an object loaded with EZbakR</span></span>
<span><span class="co"># called ode_models. More on this later</span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZDynamics.html">EZDynamics</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                    graph <span class="op">=</span> <span class="va">graph</span>,</span>
<span>                    sub_features <span class="op">=</span> <span class="st">"GF"</span>,</span>
<span>                    grouping_features <span class="op">=</span> <span class="st">"GF"</span>,</span>
<span>                    sample_feature <span class="op">=</span> <span class="st">"compartment"</span>,</span>
<span>                    modeled_to_measured <span class="op">=</span> <span class="va">ode_models</span><span class="op">$</span><span class="va">nuc2cyto</span><span class="op">$</span><span class="va">formulas</span><span class="op">)</span></span></code></pre></div>
<p>Assessing accuracy since this is simulated data with known ground
truth:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gt</span> <span class="op">&lt;-</span> <span class="va">simdata</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">$</span><span class="va">parameter_truth</span></span>
<span></span>
<span><span class="va">dynfit</span> <span class="op">&lt;-</span> <span class="va">ezbdo</span><span class="op">$</span><span class="va">dynamics</span><span class="op">$</span><span class="va">dynamics1</span> </span>
<span></span>
<span><span class="va">compare</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">dynfit</span>, <span class="va">gt</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>GF <span class="op">=</span> <span class="va">feature</span><span class="op">)</span>,</span>
<span>                             by <span class="op">=</span> <span class="st">"GF"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">true_scale_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">compare</span><span class="op">$</span><span class="va">logk1</span><span class="op">[</span><span class="va">compare</span><span class="op">$</span><span class="va">logk1</span> <span class="op">&lt;</span> <span class="fl">9.9</span><span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="va">compare</span><span class="op">$</span><span class="va">true_k1</span><span class="op">[</span><span class="va">compare</span><span class="op">$</span><span class="va">logk1</span> <span class="op">&lt;</span> <span class="fl">9.9</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">gPk1</span> <span class="op">&lt;-</span> <span class="va">compare</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>density <span class="op">=</span> <span class="fu">get_density</span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">true_k1</span><span class="op">)</span>,</span>
<span>    y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">logk1</span><span class="op">)</span><span class="op">/</span><span class="va">true_scale_factor</span><span class="op">)</span>,</span>
<span>    n <span class="op">=</span> <span class="fl">200</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">true_k1</span><span class="op">)</span>,</span>
<span>             y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">logk1</span><span class="op">)</span><span class="op">/</span><span class="va">true_scale_factor</span><span class="op">)</span>,</span>
<span>             color <span class="op">=</span> <span class="va">density</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"log(true ksyn)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"log(estimated ksyn)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span><span class="fl">1</span>,</span>
<span>              intercept <span class="op">=</span> <span class="fl">0</span>,</span>
<span>              color <span class="op">=</span> <span class="st">'darkred'</span>,</span>
<span>              linewidth <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>              linetype <span class="op">=</span> <span class="st">'dotted'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>    axis.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,</span>
<span>    legend.text<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>    legend.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span></span>
<span><span class="va">gPk2</span> <span class="op">&lt;-</span> <span class="va">compare</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>density <span class="op">=</span> <span class="fu">get_density</span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">true_k2</span><span class="op">)</span>,</span>
<span>    y <span class="op">=</span> <span class="va">logk2</span>,</span>
<span>    n <span class="op">=</span> <span class="fl">200</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">true_k2</span><span class="op">)</span>,</span>
<span>             y <span class="op">=</span> <span class="va">logk2</span>,</span>
<span>             color <span class="op">=</span> <span class="va">density</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"log(true kexp)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"log(estimated kexp)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span><span class="fl">1</span>,</span>
<span>              intercept <span class="op">=</span> <span class="fl">0</span>,</span>
<span>              color <span class="op">=</span> <span class="st">'darkred'</span>,</span>
<span>              linewidth <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>              linetype <span class="op">=</span> <span class="st">'dotted'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>    axis.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,</span>
<span>    legend.text<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>    legend.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gPk3</span> <span class="op">&lt;-</span> <span class="va">compare</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>density <span class="op">=</span> <span class="fu">get_density</span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">true_k3</span><span class="op">)</span>,</span>
<span>    y <span class="op">=</span> <span class="va">logk3</span>,</span>
<span>    n <span class="op">=</span> <span class="fl">200</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">true_k3</span><span class="op">)</span>,</span>
<span>             y <span class="op">=</span> <span class="va">logk3</span>,</span>
<span>             color <span class="op">=</span> <span class="va">density</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"log(true kdeg)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"log(estimated kdeg)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span><span class="fl">1</span>,</span>
<span>              intercept <span class="op">=</span> <span class="fl">0</span>,</span>
<span>              color <span class="op">=</span> <span class="st">'darkred'</span>,</span>
<span>              linewidth <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>              linetype <span class="op">=</span> <span class="st">'dotted'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>    axis.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,</span>
<span>    legend.text<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>    legend.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gPk1</span></span>
<span><span class="va">gPk2</span></span>
<span><span class="va">gPk3</span></span></code></pre></div>
<p><img src="EZDynamics_files/figure-html/pressure1-1.png" class="r-plt" width="40%"><img src="EZDynamics_files/figure-html/pressure1-2.png" class="r-plt" width="40%"><img src="EZDynamics_files/figure-html/pressure1-3.png" class="r-plt" width="40%"></p>
</div>
<div class="section level3">
<h3 id="modeling-nuclear-degradation">Modeling nuclear degradation<a class="anchor" aria-label="anchor" href="#modeling-nuclear-degradation"></a>
</h3>
<p>Above, we fit a model where nuclear RNA cannot degrade. What if we no
longer want to make that assumption? Just change the graph to specify
that N can degrade to 0:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Input #2: the graph</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>,</span>
<span>                  <span class="fl">3</span>, <span class="fl">0</span>, <span class="fl">2</span>,</span>
<span>                  <span class="fl">4</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                nrow <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                ncol <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"N"</span>, <span class="st">"C"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Each row and column represents an RNA species being modeled</span></span>
<span><span class="co"># Here we are modeling the dynamics of nuclear (N) and cytoplasmic (C)</span></span>
<span><span class="co"># RNA. "0" represents no RNA, and must appear in the graph. N is synthesized</span></span>
<span><span class="co"># from 0 (i.e., its synthesis kinetics are 0-th order), and C is degraded to 0</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># Numbers in graph should range from 0 to the number of parameters in your model.</span></span>
<span><span class="co"># Order is arbitrary</span></span></code></pre></div>
<p>Analyzing the data with this model is the same as before:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">##### SIMULATE DATA</span></span>
<span></span>
<span><span class="va">simdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZSimulate.html">EZSimulate</a></span><span class="op">(</span>nfeatures <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                      ntreatments <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                      mode <span class="op">=</span> <span class="st">"dynamics"</span>,</span>
<span>                      label_time <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                      dynamics_preset <span class="op">=</span> <span class="st">"nuc2cytowithNdeg"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##### EZBAKR ANALYSES</span></span>
<span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZbakRData.html">EZbakRData</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span>, <span class="va">simdata</span><span class="op">$</span><span class="va">metadf</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EstimateFractions.html">EstimateFractions</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span>
<span><span class="co">#&gt; Estimating mutation rates</span></span>
<span><span class="co">#&gt; Summarizing data for feature(s) of interest</span></span>
<span><span class="co">#&gt; Averaging out the nucleotide counts for improved efficiency</span></span>
<span><span class="co">#&gt; Estimating fractions</span></span>
<span><span class="co">#&gt; Processing output</span></span>
<span></span>
<span></span>
<span><span class="co"># Lengths of features:</span></span>
<span><span class="co"># In simulation, all features are same length, so this is here for illustrative</span></span>
<span><span class="co"># purposes only</span></span>
<span><span class="va">feature_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  GF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span><span class="op">$</span><span class="va">GF</span><span class="op">)</span>,</span>
<span>  length <span class="op">=</span> <span class="fl">1000</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Averaging fractions rather than kinetics</span></span>
<span><span class="co"># Note the model, which specifies that we want to average data for a given</span></span>
<span><span class="co"># compartment AND label time (tl).</span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AverageAndRegularize.html">AverageAndRegularize</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                              formula_mean <span class="op">=</span> <span class="op">~</span><span class="va">tl</span><span class="op">:</span><span class="va">compartment</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>                              type <span class="op">=</span> <span class="st">"fractions"</span>,</span>
<span>                              feature_lengths <span class="op">=</span> <span class="va">feature_lengths</span>,</span>
<span>                              parameter <span class="op">=</span> <span class="st">"logit_fraction_highTC"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Fitting linear model</span></span>
<span><span class="co">#&gt; Estimating coverage vs. variance trend</span></span>
<span><span class="co">#&gt; Regularizing variance estimates</span></span>
<span></span>
<span><span class="co"># Fit ODE model</span></span>
<span><span class="co"># Getting some help from ode_models object again</span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZDynamics.html">EZDynamics</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                    graph <span class="op">=</span> <span class="va">graph</span>,</span>
<span>                    sub_features <span class="op">=</span> <span class="st">"GF"</span>,</span>
<span>                    grouping_features <span class="op">=</span> <span class="st">"GF"</span>,</span>
<span>                    sample_feature <span class="op">=</span> <span class="st">"compartment"</span>,</span>
<span>                    modeled_to_measured <span class="op">=</span> <span class="va">ode_models</span><span class="op">$</span><span class="va">nuc2cytowithNdeg</span><span class="op">$</span><span class="va">formulas</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="modeling-pre--and-mature-rna-dynamics">Modeling pre- and mature RNA dynamics<a class="anchor" aria-label="anchor" href="#modeling-pre--and-mature-rna-dynamics"></a>
</h3>
<p>Data: Total RNA NR-seq data</p>
<p>Goal: Fit the simple model 0 -&gt; P -&gt; M -&gt; 0; pre-RNA (P) is
synthesized, gets preprocessed into mature RNA (M), which is eventually
degraded.</p>
<p>In this case, your cB file will need at least two feature columns. In
this example we will actually have 3, one called GF, which represents
purely intronic data, one called XF, which represents purely exonic
data, and one called feature, which is just a convenient union of XF and
GF (i.e., it will have the value of XF or GF that isn’t “__no_feature”,
meaning not assignable to that particular group). Here, we will assume
that XF reads correspond solely to mature RNA, and that GF reads
correspond solely to premature RNA. All of these choices can be altered
though.</p>
<p>The graph in this case looks just like the nuclear to cytoplasmic
model without nuclear degradation. In fact, it’s identical but with the
column and row names changed for clarity:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>,</span>
<span>                  <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">2</span>,</span>
<span>                  <span class="fl">3</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                nrow <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                ncol <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"P"</span>, <span class="st">"M"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span></span></code></pre></div>
<p>A couple things change this time when running the model:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="co">##### SIMULATE DATA</span></span>
<span></span>
<span><span class="va">simdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZSimulate.html">EZSimulate</a></span><span class="op">(</span>nfeatures <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                      ntreatments <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                      mode <span class="op">=</span> <span class="st">"dynamics"</span>,</span>
<span>                      label_time <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                      dynamics_preset <span class="op">=</span> <span class="st">"preRNA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##### EZBAKR ANALYSES</span></span>
<span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZbakRData.html">EZbakRData</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span>, <span class="va">simdata</span><span class="op">$</span><span class="va">metadf</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EstimateFractions.html">EstimateFractions</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span>
<span><span class="co">#&gt; Estimating mutation rates</span></span>
<span><span class="co">#&gt; Summarizing data for feature(s) of interest</span></span>
<span><span class="co">#&gt; Averaging out the nucleotide counts for improved efficiency</span></span>
<span><span class="co">#&gt; Estimating fractions</span></span>
<span><span class="co">#&gt; Processing output</span></span>
<span></span>
<span></span>
<span><span class="co"># Lengths of features:</span></span>
<span><span class="co"># In simulation, all features are same length, so this is here for illustrative</span></span>
<span><span class="co"># purposes only</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span><span class="op">$</span><span class="va">feature</span><span class="op">)</span></span>
<span><span class="va">feature_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  GF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">features</span>,</span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"__no_feature"</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  XF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"__no_feature"</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span><span class="op">)</span>,</span>
<span>         <span class="va">features</span><span class="op">)</span>,</span>
<span>  length <span class="op">=</span> <span class="fl">1000</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Averaging fractions rather than kinetics</span></span>
<span><span class="co"># Note the model, which specifies that we want to average data for a given</span></span>
<span><span class="co"># compartment AND label time (tl).</span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AverageAndRegularize.html">AverageAndRegularize</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                              formula_mean <span class="op">=</span> <span class="op">~</span><span class="va">tl</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>                              type <span class="op">=</span> <span class="st">"fractions"</span>,</span>
<span>                              feature_lengths <span class="op">=</span> <span class="va">feature_lengths</span>,</span>
<span>                              parameter <span class="op">=</span> <span class="st">"logit_fraction_highTC"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Fitting linear model</span></span>
<span><span class="co">#&gt; Estimating coverage vs. variance trend</span></span>
<span><span class="co">#&gt; Regularizing variance estimates</span></span>
<span></span>
<span><span class="co"># Fit ODE model</span></span>
<span><span class="co"># Getting some help from ode_models object again</span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZDynamics.html">EZDynamics</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                    graph <span class="op">=</span> <span class="va">graph</span>,</span>
<span>                    sub_features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GF"</span>, <span class="st">"XF"</span><span class="op">)</span>,</span>
<span>                    grouping_features <span class="op">=</span> <span class="st">"feature"</span>,</span>
<span>                    modeled_to_measured <span class="op">=</span> <span class="va">ode_models</span><span class="op">$</span><span class="va">preRNA</span><span class="op">$</span><span class="va">formulas</span><span class="op">$</span><span class="va">total</span><span class="op">)</span></span></code></pre></div>
<p>The main differences to note are:</p>
<ol style="list-style-type: decimal">
<li>The feature length table now has to include the set of features that
uniquely define all distinct features. In this case that means at least
GF and XF.</li>
<li>Our formula_mean only includes the label time.</li>
<li>There are two sub_features, GF and XF.</li>
<li>The feature column is the grouping_feature.</li>
<li>There is no sample_feature in EZDynamics. This is because all
modeled species (premature and mature RNA) are assayed in all
samples.</li>
</ol>
</div>
<div class="section level3">
<h3 id="cytoplasmic-to-nuclear-flow-of-pre--and-mature-rna">Cytoplasmic to nuclear flow of pre- and mature RNA<a class="anchor" aria-label="anchor" href="#cytoplasmic-to-nuclear-flow-of-pre--and-mature-rna"></a>
</h3>
<p>We can mix and match the two types of data we have discussed thus
far, namely nuclear + cytoplasmic fractionation and total RNA data. This
allows us to model the nuclear and cytoplasmic dynamics of both
premature and mature RNA. For example, consider a model where nuclear
premature RNA (NP) is first synthesized, and then either processed to
mature nuclear RNA (NM), or exported to the cytoplasm (CP). NM can then
be exported to mature cytoplasmic RNA (CM), and CP can be processed to
CM. The graph in this case looks like:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                  <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">0</span>,</span>
<span>                  <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">4</span>,</span>
<span>                  <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">5</span>,</span>
<span>                  <span class="fl">6</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                nrow <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                ncol <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"NP"</span>, <span class="st">"NM"</span>, <span class="st">"CP"</span>,<span class="st">"CM"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span></span></code></pre></div>
<p>Fitting this model looks like a combination of the nuc2cyto and
preRNA models:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##### SIMULATE DATA</span></span>
<span></span>
<span><span class="va">simdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZSimulate.html">EZSimulate</a></span><span class="op">(</span>nfeatures <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                      ntreatments <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                      mode <span class="op">=</span> <span class="st">"dynamics"</span>,</span>
<span>                      label_time <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                      dynamics_preset <span class="op">=</span> <span class="st">"nuc2cytowithpreRNA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##### EZBAKR ANALYSES</span></span>
<span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZbakRData.html">EZbakRData</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span>, <span class="va">simdata</span><span class="op">$</span><span class="va">metadf</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EstimateFractions.html">EstimateFractions</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span>
<span><span class="co">#&gt; Estimating mutation rates</span></span>
<span><span class="co">#&gt; Summarizing data for feature(s) of interest</span></span>
<span><span class="co">#&gt; Averaging out the nucleotide counts for improved efficiency</span></span>
<span><span class="co">#&gt; Estimating fractions</span></span>
<span><span class="co">#&gt; Processing output</span></span>
<span></span>
<span></span>
<span><span class="co"># Lengths of features:</span></span>
<span><span class="co"># In simulation, all features are same length, so this is here for illustrative</span></span>
<span><span class="co"># purposes only</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span><span class="op">$</span><span class="va">feature</span><span class="op">)</span></span>
<span><span class="va">feature_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  GF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">features</span>,</span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"__no_feature"</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  XF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"__no_feature"</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span><span class="op">)</span>,</span>
<span>         <span class="va">features</span><span class="op">)</span>,</span>
<span>  length <span class="op">=</span> <span class="fl">1000</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Averaging fractions rather than kinetics</span></span>
<span><span class="co"># Note the model, which specifies that we want to average data for a given</span></span>
<span><span class="co"># compartment AND label time (tl).</span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AverageAndRegularize.html">AverageAndRegularize</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                              formula_mean <span class="op">=</span> <span class="op">~</span><span class="va">tl</span><span class="op">:</span><span class="va">compartment</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>                              type <span class="op">=</span> <span class="st">"fractions"</span>,</span>
<span>                              feature_lengths <span class="op">=</span> <span class="va">feature_lengths</span>,</span>
<span>                              parameter <span class="op">=</span> <span class="st">"logit_fraction_highTC"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Fitting linear model</span></span>
<span><span class="co">#&gt; Estimating coverage vs. variance trend</span></span>
<span><span class="co">#&gt; Regularizing variance estimates</span></span>
<span></span>
<span><span class="co"># Fit ODE model</span></span>
<span><span class="co"># Getting some help from ode_models object again</span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZDynamics.html">EZDynamics</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                    graph <span class="op">=</span> <span class="va">graph</span>,</span>
<span>                    sub_features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GF"</span>, <span class="st">"XF"</span><span class="op">)</span>,</span>
<span>                    grouping_features <span class="op">=</span> <span class="st">"feature"</span>,</span>
<span>                    sample_feature <span class="op">=</span> <span class="st">"compartment"</span>,</span>
<span>                    modeled_to_measured <span class="op">=</span> <span class="va">ode_models</span><span class="op">$</span><span class="va">nuc2cytowithpreRNA</span><span class="op">$</span><span class="va">formulas</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="performing-comparative-analyses">Performing comparative analyses<a class="anchor" aria-label="anchor" href="#performing-comparative-analyses"></a>
</h3>
<p>All of the above examples involved estimating kinetic parameters in a
single biological condition. What if you have multiple distinct
conditions and want to compare parameter estimates across these
conditions? EZbakR can do that too! Here’s an example of how:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### SIMULATE DATA</span></span>
<span></span>
<span><span class="co"># Now we are simulating two distinct biological conditions.</span></span>
<span><span class="co"># Default is for about 50% of all features to have differences in</span></span>
<span><span class="co"># all parameters. </span></span>
<span><span class="va">simdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZSimulate.html">EZSimulate</a></span><span class="op">(</span>nfeatures <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                      ntreatments <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                      mode <span class="op">=</span> <span class="st">"dynamics"</span>,</span>
<span>                      label_time <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                      dynamics_preset <span class="op">=</span> <span class="st">"nuc2cyto"</span><span class="op">)</span></span>
<span>              </span>
<span><span class="co">### EZBAKR ANALYSES</span></span>
<span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZbakRData.html">EZbakRData</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span>, <span class="va">simdata</span><span class="op">$</span><span class="va">metadf</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EstimateFractions.html">EstimateFractions</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span>
<span><span class="co">#&gt; Estimating mutation rates</span></span>
<span><span class="co">#&gt; Summarizing data for feature(s) of interest</span></span>
<span><span class="co">#&gt; Averaging out the nucleotide counts for improved efficiency</span></span>
<span><span class="co">#&gt; Estimating fractions</span></span>
<span><span class="co">#&gt; Processing output</span></span>
<span></span>
<span></span>
<span><span class="co"># Lengths of features:</span></span>
<span><span class="co"># In simulation, all features are same length, so this is here for illustrative</span></span>
<span><span class="co"># purposes only</span></span>
<span><span class="va">feature_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  GF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span><span class="op">$</span><span class="va">GF</span><span class="op">)</span>,</span>
<span>  length <span class="op">=</span> <span class="fl">1000</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># DIFFERENT FROM SINGLE CONDITION ANALYSIS: "treatment" is included in the formula.</span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AverageAndRegularize.html">AverageAndRegularize</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                              formula_mean <span class="op">=</span> <span class="op">~</span><span class="va">tl</span><span class="op">:</span><span class="va">compartment</span><span class="op">:</span><span class="va">treatment</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>                              type <span class="op">=</span> <span class="st">"fractions"</span>,</span>
<span>                              feature_lengths <span class="op">=</span> <span class="va">feature_lengths</span>,</span>
<span>                              parameter <span class="op">=</span> <span class="st">"logit_fraction_highTC"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Fitting linear model</span></span>
<span><span class="co">#&gt; Estimating coverage vs. variance trend</span></span>
<span><span class="co">#&gt; Regularizing variance estimates</span></span>
<span></span>
<span></span>
<span><span class="co"># Nothing changes here!</span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZDynamics.html">EZDynamics</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                    graph <span class="op">=</span> <span class="va">ode_models</span><span class="op">$</span><span class="va">nuc2cyto</span><span class="op">$</span><span class="va">graph</span>,</span>
<span>                    sub_features <span class="op">=</span> <span class="st">"GF"</span>,</span>
<span>                    grouping_features <span class="op">=</span> <span class="st">"GF"</span>,</span>
<span>                    sample_feature <span class="op">=</span> <span class="st">"compartment"</span>,</span>
<span>                    modeled_to_measured <span class="op">=</span> <span class="va">ode_models</span><span class="op">$</span><span class="va">nuc2cyto</span><span class="op">$</span><span class="va">formulas</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Now we can run CompareParameters. Need to:</span></span>
<span><span class="co"># 1) Let it know you want to use EZDynamics output (type = "dynamics")</span></span>
<span><span class="co"># 2) Specify the "design_factor" (treatment in this case) </span></span>
<span><span class="co"># 3) Specify the "parameter" you want to compare </span></span>
<span><span class="co"># (let's compare log(k2), a.k.a logk2).</span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CompareParameters.html">CompareParameters</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                           type <span class="op">=</span> <span class="st">"dynamics"</span>,</span>
<span>                           parameter <span class="op">=</span> <span class="st">"logk2"</span>,</span>
<span>                           design_factor <span class="op">=</span> <span class="st">"treatment"</span>,</span>
<span>                           reference <span class="op">=</span> <span class="st">"treatment1"</span>,</span>
<span>                           experimental <span class="op">=</span> <span class="st">"treatment2"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Can make a volcano plot</span></span>
<span><span class="co"># Here I have specified everything to be exhaustive, but</span></span>
<span><span class="co"># in this case we could just specify `parameter = "logk2"`,</span></span>
<span><span class="co"># since we only have one logk2 comparative analysis in our</span></span>
<span><span class="co"># EZbakRData object</span></span>
<span><span class="fu"><a href="../reference/EZVolcanoPlot.html">EZVolcanoPlot</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>              parameter <span class="op">=</span> <span class="st">"logk2"</span>,</span>
<span>              design_factor <span class="op">=</span> <span class="st">"treatment"</span>,</span>
<span>              reference <span class="op">=</span> <span class="st">"treatment1"</span>,</span>
<span>              experimental <span class="op">=</span> <span class="st">"treatment2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="EZDynamics_files/figure-html/unnamed-chunk-12-1.png" class="r-plt" width="576"></p>
</div>
</div>
<div class="section level2">
<h2 id="taking-a-deeper-dive">Taking a deeper dive<a class="anchor" aria-label="anchor" href="#taking-a-deeper-dive"></a>
</h2>
<p>In this section we will explore in greater detail the many components
that make <code>EZDynamics</code> tick. We will discuss their structure
and the information they convey.</p>
<div class="section level3">
<h3 id="running-ezdynamics">Running EZDynamics<a class="anchor" aria-label="anchor" href="#running-ezdynamics"></a>
</h3>
<p><code>EZDynamics</code> has the following important/required
parameters:</p>
<ol style="list-style-type: decimal">
<li>An <code>EZbakRData</code> object on which you have run
<code><a href="../reference/AverageAndRegularize.html">AverageAndRegularize()</a></code>.</li>
<li>A matrix, referred to as the “graph” (like the mathematical concept,
not the things you make with ggplot), specifying your RNA dynamical
systems model.</li>
<li>A list of formulas describing how each measured feature relates to
the features you model in your “graph”.</li>
<li>The feature columns mentioned in your list of formulas
(<code>sub_features</code>).</li>
<li>The feature columns that will define the overarching features for
which you estimate kinetic parameters
(<code>grouping_features</code>).</li>
<li>The column in your metadf that distinguishes different samples
assying different species of RNA. (<code>sample_feature</code>).</li>
</ol>
<p>To read more about each of these, check out the function
documentation with ?EZDynamics. The “graph” and list of formulas will be
described more below.</p>
</div>
<div class="section level3">
<h3 id="the-graph">The graph<a class="anchor" aria-label="anchor" href="#the-graph"></a>
</h3>
<p>This is what tells <code>EZDynamics</code> the model you would like
to fit. It should be a matrix with as many rows as it has columns. Each
row/column represents a given specie of RNA. One of these rows/columns
must always be “0”, which will denote “no RNA”. RNA can be synthesized
from “no RNA”, and it can be degraded to “no RNA”. The other
rows/columns will be more tangible RNA species, like premature RNA, or
cytoplasmic RNA. Their names can be whatever you want, but for your
sake, try to make them make sense.</p>
<p>Element i,j (ith row, jth column) of this matrix specifies whether or
not species i can be converted into species j. If so, this element
should be an integer index from 1 to the number of possible conversions
(i.e., the number of parameters in your model). Which integer is
arbitrary, but I typically like filling out the matrix row by row (top
to bottom), from left to right.</p>
<p>For example, if we wanted to model the nuclear export of synthesized
RNA (N), and the eventual degradation of cytoplasmic RNA (C), we could
define the following “graph”:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>,</span>
<span>    <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">2</span>,</span>
<span>    <span class="fl">3</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"N"</span>, <span class="st">"C"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span></span></code></pre></div>
<p>The first row/column corresponds to “no RNA”, the second row/column
corresponds to “nuclear RNA”, and the third row/column corresponds to
“cytoplasmic RNA”. Thus:</p>
<p>Row 1, column 1 = 0 -&gt; 0, which is a trivial non-transformation,
and thus receives a value of 0. <strong>NOTE: the diagonal of this
matrix should always be 0, for this same reason</strong>.</p>
<p><strong>Row 1, column 2</strong> = 0 -&gt; N, which represents
zeroth-order synthesis of nuclear RNA, a process which we are assuming
occurs. Thus, it gets a non-zero integer index. We have chosen 1 in this
case</p>
<p><strong>Row 1, column 3</strong> = 0 -&gt; C, which would represent
zeroth-order synthesis of cytoplasmic RNA, a process we assume never
occurs. Cytoplasmic RNA in our model only comes from nuclear RNA. Thus,
it gets a 0</p>
<p><strong>Row 2, column 1</strong> = N -&gt; 0, which represents
first-order degradation of nuclear RNA, which in this case we assume
does not happen. It gets a 0.</p>
<p><strong>Row 2, column 2</strong> = N -&gt; N, trivially 0</p>
<p><strong>Row 2, column 3</strong> = N -&gt; C, which represents
first-order conversion of nuclear RNA to cytoplasmic RNA, better known
as nuclear export. We assume this does occur, and so it gets an integer
index, which we have chosen to be 2.</p>
<p><strong>Row 3, column 1</strong> = C -&gt; 0, which represents
first-order degradation of cytoplasmic RNA, which we assume occurs. This
thus gets an integer index, which we have chosen to be 3.</p>
<p><strong>Row 3, column 2</strong> = C -&gt; N, which would represent
re-entry of cytoplasmic RNA into the nucleus, which we are assuming
doesn’t happen, and thus gets a value of 0.</p>
<p><strong>Row 3, column 3</strong> = C -&gt; C, trivially 0</p>
<p>Our model has three parameters, the rate of synthesis of nuclear RNA,
the rate constant for nuclear export, and the rate constant for
cytoplasmic RNA degradation. Thus, our graph matrix needs to have a 1,
2, and 3 somewhere in it (integer indexes for each of these processes),
which it does.</p>
<div class="section level4">
<h4 id="identifiability">Identifiability<a class="anchor" aria-label="anchor" href="#identifiability"></a>
</h4>
<p>Left to your own devices, I am sure you could dream up some
fantastical and mind-bending models of RNA dynamics. Most of these
models will not be “fittable”. That is to say, you likely don’t have the
data necessary to estimate reasonable values for all of the parameters
in some of these intricate models. When this is the case, we say that
the model is not “identifiable”. Roughly speaking, a model in this case
will be identifiable if the number of parameters &lt;= 2 * (the number
of assayed RNA species). For example, in the case of our
nuclear/cytoplasmic model discussed above, there are 3 parameters.
Typically, if you are trying to fit such a model, you have nuclear
fraction NR-seq data, cytoplasmic fraction NR-seq data, and whole cell
NR-seq data. In the nuclear fraction you assay one specie of RNA
(nuclear RNA, denoted N in the graph constructed above). In the
cytoplasmic fraction you assay one specie of RNA (cytoplasmic RNA,
denoted C in the graph above). From these two alone, you have 2 species,
which means you can in theory estimate up to 4 parameters. Can you
theoretically fit a more complicated model with whole cell data? Well
your whole cell data assays the combination of two species, nuclear (N)
and cytoplasmic (C) RNA. But these were both already assayed in other
samples, thus the whole cell data doesn’t provide anything in terms of
additional model complexity freedom.</p>
<p>That is not to say that the whole cell data in this case isn’t
important. In fact, it is key to normalizing this data (see more on that
below). Without it, you would not be able to use the read counts from
each fraction to help infer kinetic parameters. In that case, the number
of parameters you could estimate would only be equal to the number of
assayed RNA species, rather than double this amount. The factor of 1
vs. 2 comes from the fact that you get two pieces of information about
each RNA specie: the fraction that is labeled, and the number of reads.
If you can’t normalize your read counts though, you effectively only get
one piece of information per specie.</p>
<p>The limit on the number of estimatable parameters quoted above is a
simple limit on the <strong>theoretical</strong> limit. It’s kind of
like a best case scenario on the number of parameters you could hope to
estimate. In practice though, not all parameters are
<strong>practically</strong> identifiable. That is to say, if you had
perfect, noise-less data you could determine their values, but with real
data, they become difficult to pin down accurately. This is often a
dynamic range problem, where a large range of values for a particular
parameter in your model is not expected to have much of an impact on the
data you observe. EZbakR can help you identify when this is the case;
see more in the section on uncertainty quantification below.</p>
</div>
</div>
<div class="section level3">
<h3 id="the-formulas">The formulas<a class="anchor" aria-label="anchor" href="#the-formulas"></a>
</h3>
<p>In addition to your model, you also need to specify to
<code><a href="../reference/EZDynamics.html">EZDynamics()</a></code> how the things you measured relate to the RNA
species you are modeling. For example, you might have a single feature
column from your cB that you analyze (e.g., XF, which in the
fastq2EZbakR output represents data for reads from definitively exonic
regions). The actual specie of RNA in your graph that this feature
represents though will differ from sample to sample though, if you are
for example analyzing subcellular fractionation NR-seq data. In
addition, there are times where you are analyzing multiple cB feature
columns (e.g., GF and XF, representing all reads from a gene, and exonic
reads from a gene, respectively). In this case, each feature will likely
represent a different RNA species, or combination of RNA species.</p>
<p>In all of these cases, formula objects are used to convey to
<code><a href="../reference/EZDynamics.html">EZDynamics()</a></code> what is what. A list of formula objects will
get passed to <code><a href="../reference/EZDynamics.html">EZDynamics()</a></code> via its
<code>modeled_to_measured</code> parameter. Each element of the list
needs to be a formula describing how the features in your cB,
potentially for a given set of samples, relates to the species in your
graph.</p>
<div class="section level4">
<h4 id="example-1-nuclear-and-cytoplasmic-fractionation">Example 1: nuclear and cytoplasmic fractionation<a class="anchor" aria-label="anchor" href="#example-1-nuclear-and-cytoplasmic-fractionation"></a>
</h4>
<p>Consider the case of modeling nuclear, cytoplasmic, and total NR-seq
data. Assume in your <code>metadf</code> you have a column called
<code>compartment</code>, which specifies which samples represent which
of these three fractions. For example, your <code>metadf</code> might
look like:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cyto_1"</span>, <span class="st">"cyto_2"</span>,</span>
<span>             <span class="st">"nuc_1"</span>, <span class="st">"nuc_2"</span>,</span>
<span>             <span class="st">"tot_1"</span>, <span class="st">"tot_3"</span><span class="op">)</span>,</span>
<span>  tl <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  compartment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cytoplasmic"</span>, <span class="st">"cytoplasmic"</span>,</span>
<span>                  <span class="st">"nuclear"</span>, <span class="st">"nuclear"</span>,</span>
<span>                  <span class="st">"total"</span>, <span class="st">"total"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Let’s also assume that your cB file has a single feature column,
let’s call it XF. The list you pass to <code>modeled_to_measured</code>
must have one element for each value of <code>compartment</code> that
shows up in your metadf. Each element needs to describe how XF in that
sample is related to the species in your graph. Say you referred to two
species in your graph, nuclear RNA (N in the graph) and cytoplasmic RNA
(C in the graph). <code>modeled_to_measured</code> would then be:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mtom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  total <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">XF</span> <span class="op">~</span> <span class="va">C</span> <span class="op">+</span> <span class="va">N</span><span class="op">)</span>,</span>
<span>  nuclear <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">XF</span> <span class="op">~</span> <span class="va">N</span><span class="op">)</span>,</span>
<span>  cytoplasmic <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">XF</span> <span class="op">~</span> <span class="va">C</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This says that:</p>
<ol style="list-style-type: decimal">
<li>In the total RNA samples, XF represents the sum of cytoplasmic (C)
and nuclear (N) RNA.</li>
<li>In the nuclear RNA samples, XF represents nuclear RNA (N).</li>
<li>In the cytoplasmic RNA samples, XF represents cytoplasmic RNA
(C).</li>
</ol>
</div>
<div class="section level4">
<h4 id="example-2-pre-rna-dynamics">Example 2: pre-RNA dynamics<a class="anchor" aria-label="anchor" href="#example-2-pre-rna-dynamics"></a>
</h4>
<p>Now, consider the case of modeling premature and mature RNA dynamics
in total NR-seq data. In this case, each sample assays the exact same
RNA species. Therefore, we won’t need to specify separate formulae for
each sample. The difference though is that your cB file will need to
have more than one feature column. For example, fastq2EZbakR can output
cB files with both XF and GF columns. XF represents data for reads
mapping to exclusively exonic regions. GF represents data for all reads
(intronic and exonic) from a gene. In this case, we will need to provide
a list with one element per feature. Here is what a reasonable
<code>modeled_to_measured</code> list might look like in this case:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mtom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">XF</span> <span class="op">~</span> <span class="va">M</span>,</span>
<span>  <span class="va">GF</span> <span class="op">~</span> <span class="va">P</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>I am assuming that M represents mature RNAs in our graph, and P
represents premature RNA. To interpret this set of formulas, we need to
note one more thing: GF will be called both a
<code>grouping_feature</code> and a <code>sub_feature</code> in our call
to <code><a href="../reference/EZDynamics.html">EZDynamics()</a></code>. As a refresher,
<code>grouping_features</code> are those which never have a value of
“__no_feature”, or some similar null-assignment value. They will be
included in the final output. <code>sub_features</code> are what you
describe in <code>modeled_to_measured</code>; they represent specific
RNA species, or combinations of RNA species.</p>
<p>In our example, because GF is both a grouping feature and a sub
feature, the GF in our <code>modeled_to_measured</code> actually
represents all entries of our cB where XF is “__no_feature” (or
similar). That is to say, it represents the case where reads overlap
intronic portions of the gene. That is why, or formulas say that:</p>
<ol style="list-style-type: decimal">
<li>XF represents the mature RNA</li>
<li>GF represents the premature RNA</li>
</ol>
<p><strong>NOTE</strong>: There is a bit of complexity here, as while
its true that intronic reads are definitively premature RNA, exonic
reads are not necessarily definitely mature RNA. It would be more
accurate to split XF into two features, call it XF_sj and XF_nosj (sj =
splice junction). The former would be instances of splice junction
mapping reads, which are (almost) definitively mature RNA, while the
latter would represent reads that don’t overlap splice junctions, and
could thus come from either premature or mature RNA. The formula in this
case would look like:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mtom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">XF_sj</span> <span class="op">~</span> <span class="va">M</span>,</span>
<span>  <span class="va">XF_nosj</span> <span class="op">~</span> <span class="va">P</span> <span class="op">+</span> <span class="va">M</span>,</span>
<span>  <span class="va">GF</span> <span class="op">~</span> <span class="va">P</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>As an aside, fastq2EZbakR includes an sj column in its output cB,
which has a value of <code>TRUE</code> for data from reads overlapping
exon-exon splice junctions, allowing you to create this XF split
yourself.</p>
</div>
<div class="section level4">
<h4 id="example-3-prernamature-rna-nuclearcytoplasmic-dynamics">Example 3: preRNA/mature RNA nuclear/cytoplasmic dynamics<a class="anchor" aria-label="anchor" href="#example-3-prernamature-rna-nuclearcytoplasmic-dynamics"></a>
</h4>
<p>What if you want to model the premature and mature RNA dynamics of
nuclear and cytoplasmic RNA? In that case, you will need to have a list
of lists, with one inner-list per compartment (e.g., total, nuclear, and
cytoplasmic) with each inner-list describing how the multiple cB feature
columns are related to the modeled species. Let’s say you are fitting
the following model (in “graph” form):</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                  <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">0</span>,</span>
<span>                  <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">4</span>,</span>
<span>                  <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">5</span>,</span>
<span>                  <span class="fl">6</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                nrow <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                ncol <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"NP"</span>, <span class="st">"NM"</span>, <span class="st">"CP"</span>,<span class="st">"CM"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span></span></code></pre></div>
<p>NP = nuclear pre-RNA; NM = nuclear mature RNA; CP = cytoplasmic
pre-RNA; and CM = cytoplasmic mature RNA. If your metadf looks like what
it did in Example 1 of this section, then
<code>modeled_to_measured</code> might look like:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mtom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  total <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="va">GF</span> <span class="op">~</span> <span class="va">NP</span> <span class="op">+</span> <span class="va">CP</span>,</span>
<span>    <span class="va">XF</span> <span class="op">~</span> <span class="va">NM</span> <span class="op">+</span> <span class="va">CM</span></span>
<span>  <span class="op">)</span>,</span>
<span>  nuclear <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="va">GF</span> <span class="op">~</span> <span class="va">NP</span>,</span>
<span>    <span class="va">XF</span> <span class="op">~</span> <span class="va">NM</span></span>
<span>  <span class="op">)</span>,</span>
<span>  cytoplasmic <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="va">GF</span> <span class="op">~</span> <span class="va">CP</span>,</span>
<span>    <span class="va">XF</span> <span class="op">~</span> <span class="va">CM</span></span>
<span>  <span class="op">)</span> </span>
<span><span class="op">)</span></span></code></pre></div>
<p>This is just a combination of the Example 1 and Example 2 lists, with
the Example 2 lists effectively inside of the Example 1 lists.</p>
</div>
</div>
<div class="section level3">
<h3 id="the-ode_models-object">The ode_models object<a class="anchor" aria-label="anchor" href="#the-ode_models-object"></a>
</h3>
<p>For your convenience, EZbakR provides a list of popular graphs and
formulas in a lazily loaded list called <code>ode_models</code>. If you
have EZbakR loaded, you can inspect the contents of this list. You can
read about its contents by running <code><a href="../reference/ode_models.html">?ode_models</a></code> in your R
session (this will pull up the data’s documentation). I’ll save you the
suspense and tell you that it has 6 elements:</p>
<ol style="list-style-type: decimal">
<li>nuc2cyto: Simplest model of nuclear and cytoplasmic RNA dynamics: 0
-&gt; N -&gt; C -&gt; 0</li>
<li>preRNA: Simplest model of pre-RNA and mature RNA dynamics: 0 -&gt; P
-&gt; M -&gt; 0</li>
<li>preRNAwithPdeg: Same as preRNA, but now pre-RNA can also
degrade.</li>
<li>nuc2cytowithNdeg: Same as nuc2cyto, but now nuclear RNA can also
degrade.</li>
<li>subtlseq: Subcellular TimeLapse-seq model, similar to that described
in <a href="https://pubmed.ncbi.nlm.nih.gov/38964322/" class="external-link">Iestwaart et al.,
2024</a>. Simplest model discussed there, lacking nuclear degradation: 0
-&gt; CH -&gt; NP -&gt; CY -&gt; PL -&gt; 0, and CY can also degrade. CH
= chromatin bound RNA. NP = nucleoplasmic RNA, not on chromatin. CY =
cytoplasmic RNA, not on polysomes. PL = polysome bound RNA.</li>
</ol>
<p>In the Quickstart documentation above, I used the formula lists
stored in these in several cases to simplify the code. They make some
hard assumptions about what you called certain compartments in your
metadf, so be a bit careful with this.</p>
</div>
<div class="section level3">
<h3 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h3>
<div class="section level4">
<h4 id="sparknotes">SparkNotes<a class="anchor" aria-label="anchor" href="#sparknotes"></a>
</h4>
<p>Below is a more detailed description of what EZbakR can do to
normalize read counts. Here, I present a brief synopsis of the main
takeaways:</p>
<ol style="list-style-type: decimal">
<li>If you want to analyze subcellular fractionation NR-seq data,
normalization is key.</li>
<li>To properly normalize subcellular fractionation data, option A for
normalization is to derive scale factors yourself from spike-ins, and
provide them via the <code>scale_factors</code> parameter. This should
be a data frame with two columns, one being “scale” (the scale factor),
and the other being whatever your <code>sample_feature</code> is (e.g.,
“compartment”). Thus, you are providing a single scale factor for all
samples from a given <code>sample_feature</code> value. Normalization
between different samples from the same <code>sample_feature</code>
value is handled by <code><a href="../reference/AverageAndRegularize.html">AverageAndRegularize()</a></code>, and is mostly
automatic. See discussion of exception below.</li>
<li>Option B for normalization is to have EZbakR figure out scale
factors. This is done via modeling of the global, length normalized
fraction of reads that are labeled in each sample. This strategy only
works if you have a sufficient set of combinations of individual sample
types. For example, if you are trying to normalize cytoplasmic/nuclear
fractionation NR-seq data, you would need cytoplasmic data, nuclear
data, but also the combination of these two (i.e., whole cell data). The
relationship between the fraction of reads labeled in these three
samples gives EZbakR the information needed to infer the relative
molecular abundances of species in each compartment.</li>
<li>If you are pursuing option B, it is important that EZbakR length
normalizes your read counts. This is done by passing a table of feature
lengths to the <code>feature_lengths</code> parameter of
<code><a href="../reference/AverageAndRegularize.html">AverageAndRegularize()</a></code>. The columns of this table should
include the names of all features you are modeling (e.g., XF), and have
an additional column called “length”. “length” should denote the
effective length of each feature.</li>
</ol>
</div>
<div class="section level4">
<h4 id="the-details">The details<a class="anchor" aria-label="anchor" href="#the-details"></a>
</h4>
<p><code><a href="../reference/EZDynamics.html">EZDynamics()</a></code> can use two pieces of information to infer
kinetic parameters in a specified model for a given feature:</p>
<ol style="list-style-type: decimal">
<li>The fraction of reads from a feature that are labeled (a.k.a., the
fraction labeled).</li>
<li>The number of reads from said feature.</li>
</ol>
<p>The former is an “internally normalized” quantity. It can be thought
of as a ratio of read counts from the same sample (# of reads from
labeled RNA / total # of reads). In RNA-seq read count normalization
scale factor applied to the numerator would be applied to the
denominator as well, and thus cancel out. Thus, the fraction labeled
does not require read count normalization to be used effectively.</p>
<p>Modeling the read counts on the other hand does require that read
counts are normalized. This usually means multiplying read counts by
sample-wide scale factors, such that differences in normalized read
count correspond to differences in absolute molecular abundances of a
given RNA. Read count normalization is difficult when the RNA
populations in different samples differ drastically, e.g. if the total
molecular abundances of RNA in one sample is greater than that in
another (say if more RNA exists in the cytoplasm than in the nucleus,
and you have data from cytoplasmic and nuclear fractions). Standard
statistical normalization strategies will fail, as they assume that no
such sample-wide differences exist. Your options for read count
normalization in the setting of analyzing subcellular fractionation data
are:</p>
<ol style="list-style-type: decimal">
<li>Derive them from spike-ins.</li>
<li>Don’t normalize read counts, and only use the fraction labeled to
infer kinetic parameters. This has the downside of greatly limiting the
complexity of models you can fit.</li>
<li>Let EZbakR infer scale factors automatically.</li>
</ol>
<p>1 is technically ideal, though getting spike-ins to work well can be
a easier said than done. If scale factors are not provided, EZbakR will
try 3. This strategy works by modeling the total fraction of reads that
are labeled in a given sample as a function of the same value in other
samples. For example, if you have cytoplasmic, nuclear, and whole cell
NR-seq data, you can model the whole cell global fraction labeled as a
weighted average of the cytoplasmic and nuclear global fraction
labeled’s. The weights are related to the relative molecular abundances
of cytoplasmic and nuclear RNA, making them precisely the normalization
scale factors we are after.</p>
</div>
</div>
<div class="section level3">
<h3 id="uncertainty-quantification">Uncertainty quantification<a class="anchor" aria-label="anchor" href="#uncertainty-quantification"></a>
</h3>
<p>The output of <code>EZDynamics</code> will be a table stored in the
<code>dynamics</code> element of your <code>EZbakRData</code> object.
You can access it with
<code>&lt;obj&gt;$dynamics$&lt;dynamics_table_name&gt;</code>, replacing
<code>&lt;obj&gt;</code> with what you called your
<code>EZbakRData</code> object, and <code>dynamics_table_name</code>
with whatever EZbakR called that table, or using <code>EZget</code>.
When interpreting the output, be mindful of the uncertainty estimates
(“se_<parameter_name>”) for a given parameter. Often, parameters in a
model are theoretically identifiable, meaning that in theory you have
enough data to estimate them, but not practically identifiable. High
uncertainties can flag instances of the latter. What represents “high”
can vary a bit from dataset to dataset and parameter to parameter, but
the range of uncertainties seen across all estimated parameters can give
you a good sense of what is good. In cases of practical
unidentifiability, there is a pretty clear stratification between low
and high uncertainty estimates.</parameter_name></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
