<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>QCing your NR-seq data with EZbakR: EZQC() • EZbakR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="QCing your NR-seq data with EZbakR: EZQC()">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">EZbakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Quickstart.html">EZbakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="../articles/EZget.html">Convenient fetching of EZbakR outputs: EZget()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateFractions.html">Universal first step: EstimateFractions()</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateKinetics.html">Standard kinetic parameter estimation: EstimateKinetics()</a></li>
    <li><a class="dropdown-item" href="../articles/Linear-modeling.html">Linear modeling in EZbakR</a></li>
    <li><a class="dropdown-item" href="../articles/EZDynamics.html">Dynamical systems modeling in EZbakR: EZDynamics()</a></li>
    <li><a class="dropdown-item" href="../articles/EZQC.html">QCing your NR-seq data with EZbakR: EZQC()</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>QCing your NR-seq data with EZbakR: EZQC()</h1>
            
      

      <div class="d-none name"><code>EZQC.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Quality control checks are an important part of any high throughput
dataset analysis, and analyzing NR-seq data is no different. Therefore,
EZbakR provides a function, <code><a href="../reference/EZQC.html">EZQC()</a></code> to help you identify
potential problems in your NR-seq data. In this section, we will discuss
how to run <code><a href="../reference/EZQC.html">EZQC()</a></code> and what it looks for. I will also
provide some suggestions for how to best design and analyze NR-seq
experiments. Let’s start by loading EZbakR, which we will use throughout
this vignette:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://isaacvock.github.io/EZbakR/">EZbakR</a></span><span class="op">)</span></span></code></pre></div>
<p><strong>NOTE</strong>: <code><a href="../reference/EZQC.html">EZQC()</a></code> is EZbakR’s instantiation
of bakR’s <code>QC_checks()</code>, though differs in a number of key
ways. Thus, don’t expect its output or behavior to exactly mimic that of
<code>QC_checks()</code>. That being said, much of the discussion of
QCing NR-seq data present in <a href="https://simonlabcode.github.io/bakR/articles/Troubleshooting.html" class="external-link">bakR’s
vignette</a> is still relevant for interpreting EZbakR’s output.</p>
</div>
<div class="section level2">
<h2 id="ezqc">EZQC<a class="anchor" aria-label="anchor" href="#ezqc"></a>
</h2>
<p><code><a href="../reference/EZQC.html">EZQC()</a></code> can take two different inputs:</p>
<ol style="list-style-type: decimal">
<li>An <code>EZbakRData</code> object. This can be created with
<code><a href="../reference/EZbakRData.html">EZbakRData()</a></code>.</li>
<li>An <code>EZbakRFractions</code> object. This is an
<code>EZbakRData</code> object on which you have run
<code><a href="../reference/EstimateFractions.html">EstimateFractions()</a></code>.</li>
</ol>
<p>An example of both of these are shown below:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZSimulate.html">EZSimulate</a></span><span class="op">(</span><span class="fl">250</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZbakRData.html">EZbakRData</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span>,</span>
<span>                    <span class="va">simdata</span><span class="op">$</span><span class="va">metadf</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Input: EZbakRData object</span></span>
<span><span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZQC.html">EZQC</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span>
<span><span class="co">#&gt; CHECKING RAW MUTATION RATES...</span></span>
<span><span class="co">#&gt; CHECKING INFERRED MUTATION RATES...</span></span>
<span><span class="co">#&gt; CHECKING READ COUNT CORRELATIONS...</span></span>
<span><span class="co">#&gt; log10(read counts) correlation for each pair of replicates are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 12 × 3</span></span></span>
<span><span class="co">#&gt;    sample_1 sample_2 correlation</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> sample1  sample2        0.984</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> sample1  sample3        0.983</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> sample1  sample7        0.983</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> sample2  sample3        0.985</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> sample2  sample7        0.986</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> sample3  sample7        0.985</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> sample4  sample5        0.982</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> sample4  sample6        0.984</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> sample4  sample8        0.986</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> sample5  sample6        0.983</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> sample5  sample8        0.985</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> sample6  sample8        0.986</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; log10(read counts) correlations are high, suggesting good reproducibility!</span></span>
<span><span class="co">#&gt; </span></span>
<span></span>
<span></span>
<span><span class="co">### Input: EZbakRFractions object</span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EstimateFractions.html">EstimateFractions</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span>
<span><span class="co">#&gt; Estimating mutation rates</span></span>
<span><span class="co">#&gt; Summarizing data for feature(s) of interest</span></span>
<span><span class="co">#&gt; Averaging out the nucleotide counts for improved efficiency</span></span>
<span><span class="co">#&gt; Estimating fractions</span></span>
<span><span class="co">#&gt; Processing output</span></span>
<span><span class="va">qc_fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZQC.html">EZQC</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span>
<span><span class="co">#&gt; CHECKING RAW MUTATION RATES...</span></span>
<span><span class="co">#&gt; CHECKING INFERRED MUTATION RATES...</span></span>
<span><span class="co">#&gt; CHECKING READ COUNT CORRELATIONS...</span></span>
<span><span class="co">#&gt; log10(read counts) correlation for each pair of replicates are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 12 × 3</span></span></span>
<span><span class="co">#&gt;    sample_1 sample_2 correlation</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> sample1  sample2        0.984</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> sample1  sample3        0.983</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> sample1  sample7        0.983</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> sample2  sample3        0.985</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> sample2  sample7        0.986</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> sample3  sample7        0.985</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> sample4  sample5        0.982</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> sample4  sample6        0.984</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> sample4  sample8        0.986</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> sample5  sample6        0.983</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> sample5  sample8        0.985</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> sample6  sample8        0.986</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; log10(read counts) correlations are high, suggesting good reproducibility!</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; CHECKING FRACTION LABELED DISTRIBUTIONS...</span></span>
<span><span class="co">#&gt; Average fractions for each sample are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt;   sample  avg_fraction fraction_type  </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample1        0.303 fraction_highTC</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sample2        0.306 fraction_highTC</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sample3        0.303 fraction_highTC</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sample4        0.309 fraction_highTC</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sample5        0.308 fraction_highTC</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> sample6        0.307 fraction_highTC</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Labeling rates (e.g., fraction labeled for single label experiments) look good!</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; CHECKING FRACTION LABELED CORRELATIONS...</span></span>
<span><span class="co">#&gt; logit(fraction_highTC) correlation for each pair of replicates are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt;   sample_1 sample_2 correlation</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample1  sample2        0.967</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sample1  sample3        0.960</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sample2  sample3        0.964</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sample4  sample5        0.950</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sample4  sample6        0.959</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> sample5  sample6        0.962</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; logit(fraction_highTC) correlations are high, suggesting good reproducibility!</span></span>
<span><span class="co">#&gt; </span></span></code></pre></div>
<p>In the first case (<code>EZbakRData</code> input), the following are
checked:</p>
<ol style="list-style-type: decimal">
<li>Correlation of read counts between replicates</li>
<li>Raw mutation rates</li>
<li>Labeled and unlabeled read mutation rates</li>
</ol>
<p>In the second case (<code>EZbakRFractions</code> input), all of the
same things are checked, with the addition of:</p>
<ol style="list-style-type: decimal">
<li>Correlation of fraction estimates (e.g., fraction labeled in single
label experiments) between replicates.</li>
<li>Distribution of fraction estimates.</li>
</ol>
<p>Replicate correlation is an intuitive QC metric that ensures
replicates agree with each other well. The other three metrics are
NR-seq specific metrics that assess the extent to which metabolic label
was readily incorporated into nascent RNA, and the appropriateness of
the metabolic label feed time used (i.e., how long cells were fed with
the label, sometimes referred to as the label time). In general, you
want to see:</p>
<ol style="list-style-type: decimal">
<li>High labeled read mutation rates (referred to as pnew within
EZbakR).</li>
<li>Low unlabeled read mutation rates (referred to as pold within
EZbakR).</li>
<li>Roughly equal proportions of all mutational populations (e.g.,
labeled and unlabeled reads in a single label experiment).</li>
</ol>
<div class="section level3">
<h3 id="ezqc-output">EZQC output<a class="anchor" aria-label="anchor" href="#ezqc-output"></a>
</h3>
<p>Inside of the output objects (<code>qc</code> and <code>qc_fn</code>
in the above code), you will find a number of plots. The output is a
named list, with one or more plot present in each element of this list.
The named elements and their contents are as follows:</p>
<ol style="list-style-type: decimal">
<li>
<code>Raw_mutrates</code>: If you have a single mutation type in
your cB, this will be a barplot of raw mutation rates of that type in
all samples. If you have multiple mutation types, then this will be a
named list, with one element per mutation type, with each element being
this same barplot.
<ul>
<li>The raw mutation rate = (total # of mutations) / (total # of mutable
nucleotides)</li>
<li>Raw mutation rate should be high in samples fed the relevant label,
and low in unfed samples.</li>
</ul>
</li>
<li>
<code>Inferred_mutrates</code>: If you have a single mutation type
in your cB, this will be a barplot of the inferred labeled and unlabeled
read mutation rates in all samples. If you have multiple mutation types,
then it will be a named list of similar barplots for each mutation type,
as in <code>Raw_mutrates</code>.
<ul>
<li>The inferred labeled mutation rate is the rate of a mutation of a
particular type seen in reads from RNA synthesized in the presence of
the relevant metabolic label. The inferred unlabeled mutation rate is
this same mutation rate, but in reads from RNA synthesized prior to
labeling. These are sometimes referred to as the conversion and
background rates, respectively.</li>
<li>The plot includes a subtitle with some guidelines for
interpretation. Inferred labeled mutation rates is preferably much
higher than the unlabeled mutation rates in all samples.</li>
</ul>
</li>
<li>
<code>Readcount_corr</code>: This is a list, whose elements are
collections of read count correlation plots for a given group of
replicates. Groups of replicates are determined by the provided metadf
in your <code>EZbakRData</code> object. +label and -label replicates of
a given treatment condition are grouped together. Different label times
for the same treatment condition are also grouped together here.
<ul>
<li>The plots are named according to the samples compared (e.g.,
“sample1_vs_sample2”).</li>
</ul>
</li>
<li>
<code>Fraction_labeled_dist</code>: This is a list of density plots
for each label fed sample. In the case of a single label experiment,
these will be the distribution of estimated fraction labeled’s for each
sample. If you have multiple labels/mutation types, then this will be a
list with density plots for each estimated fraction.
<ul>
<li>Ideally, these will peak around 1 / (# of populations). For a
standard, single label experiment, this would be 0.5 (# of populations =
2; labeled and unlabeled).</li>
<li>This is only present if you provided a <code>EZbakRFractions</code>
object as input!</li>
</ul>
</li>
<li>
<code>Fraction_labeled_corr</code>: Same as
<code>Readcount_corr</code>, but this time correlating fraction
estimates provided by <code><a href="../reference/EstimateFractions.html">EstimateFractions()</a></code>. Unlike in
<code>Readcount_corr</code>, this excludes -label samples, and considers
distinct label times distinct replicates.
<ul>
<li>This is only present if you provided a <code>EZbakRFractions</code>
object as input!</li>
</ul>
</li>
</ol>
</div>
<div class="section level3">
<h3 id="what-to-do-if-the-qc-looks-iffy">What to do if the QC looks iffy?<a class="anchor" aria-label="anchor" href="#what-to-do-if-the-qc-looks-iffy"></a>
</h3>
<p><strong>Potential problem #1</strong>: Labeled read mutation rates
are low</p>
<p>Possible solutions:</p>
<ol style="list-style-type: decimal">
<li>Better read alignment settings to ensure recovery of high mutation
content reads. <a href="https://github.com/isaacvock/fastq2EZbakR" class="external-link">fastq2EZbakR’s</a>
default <a href="https://github.com/isaacvock/fastq2EZbakR/blob/main/config/config.yaml" class="external-link">config</a>
includes STAR settings used by multiple labs to help with this. Tools
like <a href="https://github.com/erhard-lab/grandRescue" class="external-link">grandRescue</a>
developed by the Erhard lab may also help.</li>
<li>If you ran fastq2EZbakR, did you properly set the strandedness of
your library? You can check some of the intermediate files produced by
fastq2EZbakR to test this hypothesis; see <a href="https://fastq2ezbakr.readthedocs.io/en/latest/faq/" class="external-link">here</a> for
details.</li>
<li>If using a short label time, or if incorporation rates are low,
EZbakR may just be having a tough time estimating the labeled read
mutation rate. If you have -label data, you can use this to infer a
single background mutation rate for all samples, and then estimate the
labeled read mutation rate, fixing this unlabeled read mutation rate.
This can be done by setting <code>pold_from_no_label=TRUE</code> in your
call to <code><a href="../reference/EstimateFractions.html">EstimateFractions()</a></code>. In fact, I even like
defaulting to this setting when -label samples are available.</li>
<li>You may need to optimize the concentration of metabolic label being
fed to cells. A simple TAMRA-labeling dot blot assay can be used to
broadly access the incorporation of metabolic label in your particular
system.</li>
</ol>
<p><strong>Potential problem #2</strong>: Labeled read mutation rates
are acceptable, but raw mutation rates are low. In this case, fraction
estimate replicate correlation may also be low.</p>
<p>Possible solutions:</p>
<ol style="list-style-type: decimal">
<li>You may need to use a longer label time to ensure better
representation of labeled RNA in your samples.</li>
<li>Either way, using <code>pold_from_no_label=TRUE</code> in your
<code><a href="../reference/EstimateFractions.html">EstimateFractions()</a></code> call, as described above, is likely a
good idea to ensure accurate labeled read mutation rate inference.</li>
</ol>
<p><strong>Potential problem #3</strong>: Poor correlation of read
counts across replicates, especially between -label and +label, or
different label time +label, samples.</p>
<p>This may be a sign of dropout, described <a href="https://academic.oup.com/nar/article/52/7/e35/7612100" class="external-link">here</a>, ,
<a href="https://www.biorxiv.org/content/10.1101/2023.05.24.542133v1" class="external-link">here</a>,
and <a href="https://simonlabcode.github.io/bakR/articles/Dropout.html" class="external-link">here</a>.
This is when reads from highly labeled RNA are underrepresented in
+label data. You can try:</p>
<ol style="list-style-type: decimal">
<li>Decreasing metabolic label concentration.</li>
<li>Decreasing metabolic label feed time.</li>
<li>Using the RNA handling protocol suggested <a href="https://www.biorxiv.org/content/10.1101/2023.05.24.542133v1" class="external-link">here</a>.</li>
<li>Better read alignment settings to avoid loss of high mutation
content reads. See the potential solutions to problem #1 described above
for details.</li>
<li>Running EZbakR’s <code><a href="../reference/CorrectDropout.html">CorrectDropout()</a></code> to try and
bioinformatically correct for the biases induced by dropout. This
requires you to have -label data from all distinct biological
conditions.</li>
</ol>
<p>Suggestions 1 and 2 can help no matter the cause of dropout (e.g.,
adverse effects of the label, RNA handling problems, read alignment
problems, etc.). The other three suggestions are more suited in the case
when dropout does not seem to have a biological origin. A general,
proven strategy to distinguish the cause of dropout does not exist, but
you may want to try assessing trends in sequencing tracks (e.g., look
dropoff in coverage near the 3’ end of transcripts upon increased
labeling) or performing differential expression and GO analysis of + and
-label samples to assess potential upregulation of transcriptional
repression stress pathways.</p>
</div>
</div>
<div class="section level2">
<h2 id="characteristics-of-a-good-nr-seq-experiment">Characteristics of a good NR-seq experiment:<a class="anchor" aria-label="anchor" href="#characteristics-of-a-good-nr-seq-experiment"></a>
</h2>
<p>Below I will list and discuss what, in my opinion, makes an ideal
NR-seq dataset:</p>
<ol style="list-style-type: decimal">
<li>Multiple +label (e.g., s4U) replicates for each biological condition
being assessed. Nothing radical here, replicates are important in all
settings.</li>
<li>At least one, though preferably multiple, replicates of -label
(e.g., regular RNA-seq) controls in each biological condition being
assessed. This is a crucial control to ensure that the label(s) used do
not have a significant impact on the treated cells. These can also help
correct for certain instances of bias (i.e., <a href="https://simonlabcode.github.io/bakR/articles/Dropout.html" class="external-link">dropout</a>)
present in some NR-seq datasets.</li>
<li>A pulse-label (label for certain amount of time and extract RNA)
rather than a pulse-chase (label for certain amount of time, then chase
with regular nucleotide for a certain amount of time, then extract RNA)
design. I’ve ranted about this <a href="https://simonlabcode.github.io/bakR/articles/Troubleshooting.html" class="external-link">elsewhere</a>
(bottom of linked vignette), but in short, a pulse-label design is
almost always optimal relative to a pulse-chase design. Pulse-chases:
<ul>
<li>suffer from longer exposure of cells to metabolic label, thus
incurring a greater risk for physiological impact.</li>
<li>require comparisons across samples to infer kinetic parameters,
yielding noisier estimates.</li>
<li>Are almost never necessary as the dynamics of the unlabeled RNA in a
pulse-label experiment are identical to that of the labeled RNA in a
pulse-chase experiment.</li>
<li>Rare exceptions to these rules exist, but unless you have a really
good reason to use a pulse-chase design, pulse-labels should be your
default.</li>
</ul>
</li>
<li>A metabolic label feed time length around that of the average
half-life of the RNAs of interest. Shorter is better than longer to
limit any adverse effects of the metabolic label.
<ul>
<li>In humans and mice, this means about 4 hours, if there isn’t a
specific species of RNA you are interested in probing. 2 hour label
times are typical and tend to work great in these systems.</li>
<li>In yeast this is much shorter, around 30 minutes.</li>
</ul>
</li>
<li>An optimized concentration of metabolic label.
<ul>
<li>Typical values range from 100
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi mathvariant="normal">μ</mi><annotation encoding="application/x-tex">\mathrm{\mu}</annotation></semantics></math>M
to 1 mM, but simple, low throughput assays (e.g., dot blots) should be
used to ensure sufficient incorporation of metabolic label at a given
concentration. Different cell lines can exhibit wildly different levels
of label uptake, and thus it would be best to assess this in each unique
biological context you plan to do NR-seq in.</li>
<li>Lower concentrations are better to avoid adverse effects, but make
sure that incorporation at lower concentrations is still robust before
committing.</li>
</ul>
</li>
</ol>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
