<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>EZbakR for people in a hurry • EZbakR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="EZbakR for people in a hurry">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">EZbakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Quickstart.html">EZbakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="../articles/EZget.html">EZget</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateFractions.html">EstimateFractions</a></li>
    <li><a class="dropdown-item" href="../articles/EstimateKinetics.html">EstimateKinetics</a></li>
    <li><a class="dropdown-item" href="../articles/Linear-modeling.html">Linear modeling in EZbakR</a></li>
    <li><a class="dropdown-item" href="../articles/EZDynamics.html">Dynamical systems modeling in EZbakR</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>EZbakR for people in a hurry</h1>
            
      

      <div class="d-none name"><code>Quickstart.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>EZbakR is a one-stop-shop for analyses of nucleotide recoding
      RNA-seq datasets (NR-seq). NR-seq refers to a class of methods
      (e.g., TimeLapse-seq, SLAM-seq, TUC-seq, etc.) that combine
      RNA-seq, metabolic labeling, and unique metabolic label recoding
      chemistries. These methods were originally developed to dissect
      the kinetics of RNA synthesis and degradation. Excitingly though,
      a treasure trove of extensions of the original methods have been
      created over the years. To-date, nucleotide recoding has been
      combined with TT-seq, Start-seq, Ribo-seq, scRNA-seq, Perturb-seq,
      and subcellular fractionation, just to name a few such extensions.
      In addition, while the original methods used 4-thiouridine
      (s<sup>4</sup>U), the same chemistry has been found to work with
      6-thioguanosine (s<sup>6</sup>G), opening the door to
      dual-labeling experimental designs (e.g., TILAC). EZbakR was
      designed to facilitate anlayses of the full gamut of NR-seq
      methods. While not all analyses implemented in EZbakR can be
      applied to all variants of NR-seq, EZbakR’s modular design means
      that it has something for everyone. In this vignette, we will
      quickly highlight the main modules of EZbakR. For simplicity, we
      will be focusing on how to perform a standard, single-label NR-seq
      analysis. Other vignettes go into greater detail about each module
      and the broad array of analyses that can be performed by each.</p>
    </div>
    
<div class="section level2">
<h2 id="necessary-setup">Necessary Setup<a class="anchor" aria-label="anchor" href="#necessary-setup"></a>
</h2>
<p>You will need to install and load EZbakR. Instructions to do so can
be found at this <a href="https://github.com/isaacvock/EZbakR" class="external-link">link</a>.
Setting the seed will also ensure exact reproducibility of analyses of
simulated data shown in this vignette. Finally, I will use a little bit
of dplyr to showcase some key results:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://isaacvock.github.io/EZbakR/">EZbakR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-0-create-an-ezbakrdata-object">Step 0: Create an EZbakRData Object<a class="anchor" aria-label="anchor" href="#step-0-create-an-ezbakrdata-object"></a>
</h2>
<p>To use EZbakR, you will need to create an <code>EZbakRData</code>
object. An <code>EZbakRData</code> object consists of two components: a
cB data frame and a metadf data frame. cB stands for “counts binomial”
and contains all of the information about mutations seen in sequencing
reads in each sample sequenced. metadf stands for “metadata data frame”
and contains information about the experimental details of each sample
(i.e., how long the metabolic label feed was). Examples of what these
data structures look like are lazily loaded when you load
<code>EZbakR</code>. This means that you can call them as if you created
them yourselves, with <code>example_cB</code> and
<code>example_metadf</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check out examples:</span></span>
<span><span class="va">example_cB</span></span>
<span><span class="va">example_metadf</span></span></code></pre></div>
<p>A cB data frame consists of rows corresponding to groups of reads
with identical data, where data corresponds to the:</p>
<ol style="list-style-type: decimal">
<li>Sample of origin. This should be noted in the
<strong>sample</strong> column of the cB.</li>
<li>Mutational content. This should be noted by one or more columns of
the form <reference base><sequenced base> (e.g., <strong>TC</strong> in
<code>example_cB</code>). The value in these columns should correspond
to the observed number of mutations.</sequenced></reference>
</li>
<li>Mutable nucleotide content. This should be noted by one or more
columns of the form n<reference base> (e.g., <strong>nT</strong> in
<code>example_cB</code>). The value in these columns should correspond
to the number of mutable bases overlapped by the read.</reference>
</li>
<li>Genomic features the read was assigned to. In EZbakR, these can be
named anything. The fastq2EZbakR pipeline most commonly paired with
EZbakR can include feature columns such as those which show up in
<code>example_cB</code>:
<ul>
<li>
<strong>GF</strong>: Stands for “gene feature” and represents the
gene (exonic or intronic regions) the read overlapped</li>
<li>
<strong>XF</strong>: Stands for “exonic feature” and represents the
gene (exonic regions only) the read overlapped</li>
<li>
<strong>rname</strong>: Chromosome the read came from.</li>
</ul>
</li>
</ol>
<p>The final column of the cB data frame is <strong>n</strong>, the
number of reads with identical data for the other columns. cB data
frames are most easily obtained from the Snakemake pipeline
fastq2EZbakR, available <a href="https://github.com/isaacvock/fastq2EZbakR" class="external-link">here</a>.</p>
<p>A metadf table has the following required columns:</p>
<ol style="list-style-type: decimal">
<li>
<strong>sample</strong>: Sample name, as it shows up in the cB
table.</li>
<li>Metabolic label time information. If you have a single label, this
should be documented in a column called <strong>tl</strong>. If you used
multiple labels, then the different label times should be noted in a
column called <strong>tl_<muttype></muttype></strong>, where <muttype> denotes the
type of mutation that conversion of the label yields. For example, if
you are using s<sup>4</sup>U (conversion yields T-to-C mutations) and
s<sup>6</sup>G (conversion yields G-to-A mutations), then your metadf
should include columns <strong>tl_TC</strong> and
<strong>tl_GA</strong>.</muttype>
</li>
</ol>
<p>The remaining columns can be named anything you want and should
describe aspects of each experiment. NOTE: these are currently
exclusively interpreted as discrete factors. Things like time-series
modeling is not currently supported, but may be in future releases.
These factors will be used when deciding how to group samples when
averaging replicate data and performing comparative analyses. More on
this later!</p>
<p>Once you have a cB and metadf table, you can create an EZbakRData
object:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZbakRData.html">EZbakRData</a></span><span class="op">(</span><span class="va">example_cB</span>, <span class="va">example_metadf</span><span class="op">)</span></span></code></pre></div>
<p>Now, let’s generate some simulated data to be used in the remaining
steps:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate 500 "genes" worth of data</span></span>
<span>  <span class="co"># Simulates 2 replicates of 2 different experimental conditions</span></span>
<span><span class="va">simdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZSimulate.html">EZSimulate</a></span><span class="op">(</span>nfeatures <span class="op">=</span> <span class="fl">500</span>, nreps <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZbakRData.html">EZbakRData</a></span><span class="op">(</span><span class="va">simdata</span><span class="op">$</span><span class="va">cB</span>, <span class="va">simdata</span><span class="op">$</span><span class="va">metadf</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-1-estimatefractions">Step 1: EstimateFractions()<a class="anchor" aria-label="anchor" href="#step-1-estimatefractions"></a>
</h2>
<p>The first step of any NR-seq analysis is to figure out what fraction
of your reads come from each expected mutational population. For
example, in a standard s<sup>4</sup>U labeling NR-seq experiment, you
will have reads from a high T-to-C mutation population and reads from a
low T-to-C mutation population. Technically, this involves two steps: 1)
estimating mutation rates of the two populations in each labeled sample
and 2) estimating fractions. Both of these steps are performed with a
single function in EZbakR: <code><a href="../reference/EZbakRFractions.html">EZbakRFractions()</a></code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EstimateFractions.html">EstimateFractions</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span>
<span><span class="co">#&gt; Estimating mutation rates</span></span>
<span><span class="co">#&gt; Summarizing data for feature(s) of interest</span></span>
<span><span class="co">#&gt; Averaging out the nucleotide counts for improved efficiency</span></span>
<span><span class="co">#&gt; Estimating fractions</span></span>
<span><span class="co">#&gt; Processing output</span></span></code></pre></div>
<p>See <code>?EstimateFractions()</code> for details and descriptions of
optional parameters. We can use the simulated ground truth to assess
estimate accuracy and confirm that the method is working:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulated ground truth</span></span>
<span><span class="va">truth</span> <span class="op">&lt;-</span> <span class="va">simdata</span><span class="op">$</span><span class="va">PerRepTruth</span></span>
<span></span>
<span><span class="co"># Get fraction estimates. </span></span>
<span>  <span class="co"># Can also access with ezbdo$fractions[[1]] or ezbdo$fractions$feature</span></span>
<span><span class="va">est</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZget.html">EZget</a></span><span class="op">(</span><span class="va">ezbdo</span>, type <span class="op">=</span> <span class="st">'fractions'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine</span></span>
<span><span class="va">compare</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">est</span>, <span class="va">truth</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'sample'</span>, <span class="st">'feature'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assess accuracy</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">compare</span><span class="op">$</span><span class="va">true_fraction_highTC</span>, <span class="va">compare</span><span class="op">$</span><span class="va">fraction_highTC</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Quickstart_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Looks good!</p>
</div>
<div class="section level2">
<h2 id="step-2-estimatekinetics">Step 2: EstimateKinetics()<a class="anchor" aria-label="anchor" href="#step-2-estimatekinetics"></a>
</h2>
<p>The next step of a standard NR-seq analysis is to use the estimates
of mutational population proportions (and possibly normalized read
counts) to estimate kinetic parameters of interest (e.g., degradation
and synthesis rate constants). The simplest and quickest way to do this
is with <code><a href="../reference/EstimateKinetics.html">EstimateKinetics()</a></code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EstimateKinetics.html">EstimateKinetics</a></span><span class="op">(</span><span class="va">ezbdo</span><span class="op">)</span></span></code></pre></div>
<p>By default, this uses a standard, steady-state, single rate constant
model of RNA metabolism to infer a degradation rate constant
(<code>kdeg</code>) and synthesis rate constant (<code>ksyn</code>) for
each gene. Other analysis strategies implemented in
<code><a href="../reference/EstimateKinetics.html">EstimateKinetics()</a></code> include those more appropriate for
non-steady-state systems or short label times. The <code>strategy</code>
argument in <code><a href="../reference/EstimateKinetics.html">EstimateKinetics()</a></code> sets which analysis strategy
is used. See <code><a href="../reference/EstimateKinetics.html">?EstimateKinetics</a></code> for details.</p>
<p>Again, let’s compare to ground_truth:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulated ground truth</span></span>
<span><span class="va">truth</span> <span class="op">&lt;-</span> <span class="va">simdata</span><span class="op">$</span><span class="va">PerRepTruth</span></span>
<span></span>
<span><span class="co"># Get fraction estimates. </span></span>
<span>  <span class="co"># Can also access with ezbdo$fractions[[1]] or ezbdo$fractions$feature</span></span>
<span><span class="va">est</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZget.html">EZget</a></span><span class="op">(</span><span class="va">ezbdo</span>, type <span class="op">=</span> <span class="st">'kinetics'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine</span></span>
<span><span class="va">compare</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">est</span>, <span class="va">truth</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'sample'</span>, <span class="st">'feature'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assess accuracy</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">compare</span><span class="op">$</span><span class="va">true_kdeg</span><span class="op">)</span>, <span class="va">compare</span><span class="op">$</span><span class="va">log_kdeg</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Quickstart_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Great!</p>
</div>
<div class="section level2">
<h2 id="step-3-averageandregularize">Step 3: AverageAndRegularize()<a class="anchor" aria-label="anchor" href="#step-3-averageandregularize"></a>
</h2>
<p>Typically, you will have multiple replicates of each experimental
condition. In that case, the next step is to average the data across
these replicates and estimate overall uncertainties. This can then be
passed along to downstream comparative analyses to see which features
have differences in their kinetic parameters between different
experimental conditions. This averaging process is done with
<code><a href="../reference/AverageAndRegularize.html">AverageAndRegularize()</a></code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AverageAndRegularize.html">AverageAndRegularize</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                              parameter <span class="op">=</span> <span class="st">"log_kdeg"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Fitting linear model</span></span>
<span><span class="co">#&gt; Estimating coverage vs. variance trend</span></span>
<span><span class="co">#&gt; Regularizing variance estimates</span></span></code></pre></div>
<p>See <code>?AverageAndRegularize()</code> for details. One cool thing
about <code><a href="../reference/AverageAndRegularize.html">AverageAndRegularize()</a></code> is that it implements a
heteroskedastic linear model and can thus be passed formula objects to
detail how you would like to stratify samples, and whether or not there
are any batches you would like to regress out of the main parameter
effects. See the relevant vignette for more details.</p>
<p>Finally, setting <code>parameter = "log_kdeg"</code> was not
technically necessary as it is the default, but it showcases that you
need to specify which kinetic parameter you want to average across
replicates. If you run this again with a different parameter, a new
table will be added to the <code>ezbdo$averages</code> list:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AverageAndRegularize.html">AverageAndRegularize</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                              parameter <span class="op">=</span> <span class="st">"log_ksyn"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Fitting linear model</span></span>
<span><span class="co">#&gt; Estimating coverage vs. variance trend</span></span>
<span><span class="co">#&gt; Regularizing variance estimates</span></span></code></pre></div>
<p>Let’s make sure the estimates are still accurate:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">truth</span> <span class="op">&lt;-</span> <span class="va">simdata</span><span class="op">$</span><span class="va">AvgTruth</span></span>
<span></span>
<span><span class="va">est</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EZget.html">EZget</a></span><span class="op">(</span><span class="va">ezbdo</span>, type <span class="op">=</span> <span class="st">'averages'</span>, parameter <span class="op">=</span> <span class="st">'log_kdeg'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">compare</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">truth</span>, <span class="va">est</span>, by <span class="op">=</span> <span class="st">'feature'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">compare</span><span class="op">$</span><span class="va">true_logkdeg_treatmenttreatment1</span>,</span>
<span>     <span class="va">compare</span><span class="op">$</span><span class="va">mean_treatmenttreatment1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Quickstart_files/figure-html/unnamed-chunk-11-1.png" width="576" style="display: block; margin: auto;"></p>
<p>All good!</p>
</div>
<div class="section level2">
<h2 id="step-4-compareparameters">Step 4: CompareParameters()<a class="anchor" aria-label="anchor" href="#step-4-compareparameters"></a>
</h2>
<p>Once you have replicate averaged parameter estimates, you can compare
them and assess the “statistical significance” of any observed
differences. This is done with <code><a href="../reference/CompareParameters.html">CompareParameters()</a></code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ezbdo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CompareParameters.html">CompareParameters</a></span><span class="op">(</span><span class="va">ezbdo</span>,</span>
<span>                           condition <span class="op">=</span> <span class="st">'treatment'</span>,</span>
<span>                           reference <span class="op">=</span> <span class="st">'treatment1'</span>,</span>
<span>                           experimental <span class="op">=</span> <span class="st">'treatment2'</span>,</span>
<span>                           parameter <span class="op">=</span> <span class="st">"log_kdeg"</span><span class="op">)</span></span></code></pre></div>
<p>In addition to providing the <code>EZbakRData</code> object output by
<code><a href="../reference/AverageAndRegularize.html">AverageAndRegularize()</a></code>, you also need to specify three
other arguments:</p>
<ol style="list-style-type: decimal">
<li>
<code>condition</code>: the metadf column (or function of columns)
that replicates were stratified by. The simulated metadf has a single
experimental detail column called <code>treatment</code>, so this is the
condition here.</li>
<li>
<code>reference</code>: what value of <code>condition</code> do you
want to consider the “reference” level.</li>
<li>
<code>experimental</code>: what value of <code>condition</code> do
you want to consider the “experimental” level.</li>
</ol>
<p>Once again, <code>parameter</code>’s default value is “log_kdeg”, but
is specified here to be explicit.<code><a href="../reference/CompareParameters.html">CompareParameters()</a></code> will
calculate the difference in average parameter as:
<code>experimental</code> - <code>reference</code>.</p>
<p>EZbakR provides some convenient visualization functions to inspect
the output of <code><a href="../reference/CompareParameters.html">CompareParameters()</a></code>.</p>
<p>Volcano plot:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/EZVolcanoPlot.html">EZVolcanoPlot</a></span><span class="op">(</span><span class="va">ezbdo</span>, parameter <span class="op">=</span> <span class="st">"log_kdeg"</span>,</span>
<span>              condition <span class="op">=</span> <span class="st">'treatment'</span>,</span>
<span>              reference <span class="op">=</span> <span class="st">'treatment1'</span>,</span>
<span>              experimental <span class="op">=</span> <span class="st">'treatment2'</span><span class="op">)</span></span></code></pre></div>
<p><img src="Quickstart_files/figure-html/unnamed-chunk-13-1.png" width="576" style="display: block; margin: auto;"></p>
<p>MA plot:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/EZMAPlot.html">EZMAPlot</a></span><span class="op">(</span><span class="va">ezbdo</span>, parameter <span class="op">=</span> <span class="st">"log_kdeg"</span>,</span>
<span>         condition <span class="op">=</span> <span class="st">'treatment'</span>,</span>
<span>         reference <span class="op">=</span> <span class="st">'treatment1'</span>,</span>
<span>         experimental <span class="op">=</span> <span class="st">'treatment2'</span><span class="op">)</span></span></code></pre></div>
<p><img src="Quickstart_files/figure-html/unnamed-chunk-14-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
